// @uipath/uipath-typescript/dist/traces/index.mjs
import { getOrCreateClient, createTrack, createTrackEvent } from '@uipath/core-telemetry';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise, SuppressedError, Symbol, Iterator */


function __decorate(decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
}

function __classPrivateFieldGet(receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
}

function __classPrivateFieldSet(receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
}

typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
    var e = new Error(message);
    return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

/**
 * Type guards for error response types
 */
function isOrchestratorError(error) {
    return typeof error === 'object' &&
        error !== null &&
        'message' in error &&
        'errorCode' in error &&
        typeof error.message === 'string' &&
        typeof error.errorCode === 'number';
}
function isEntityError(error) {
    return typeof error === 'object' &&
        error !== null &&
        'error' in error &&
        typeof error.error === 'string';
}
function isPimsError(error) {
    return typeof error === 'object' &&
        error !== null &&
        'type' in error &&
        'title' in error &&
        'status' in error &&
        typeof error.type === 'string' &&
        typeof error.title === 'string' &&
        typeof error.status === 'number';
}

/**
 * HTTP status code constants for error handling
 */
const HttpStatus = {
    // Client errors (4xx)
    BAD_REQUEST: 400,
    UNAUTHORIZED: 401,
    FORBIDDEN: 403,
    NOT_FOUND: 404,
    TOO_MANY_REQUESTS: 429,
    // Server errors (5xx)
    INTERNAL_SERVER_ERROR: 500,
    BAD_GATEWAY: 502,
    SERVICE_UNAVAILABLE: 503,
    GATEWAY_TIMEOUT: 504
};
/**
 * Error type constants for consistent error identification
 */
const ErrorType = {
    AUTHENTICATION: 'AuthenticationError',
    AUTHORIZATION: 'AuthorizationError',
    VALIDATION: 'ValidationError',
    NOT_FOUND: 'NotFoundError',
    RATE_LIMIT: 'RateLimitError',
    SERVER: 'ServerError',
    NETWORK: 'NetworkError'
};
/**
 * HTTP header constants for error handling
 */
const HttpHeaders = {
    X_REQUEST_ID: 'x-request-id'
};
/**
 * Standard error message constants
 */
const ErrorMessages = {
    // Authentication errors
    AUTHENTICATION_FAILED: 'Authentication failed',
    // Authorization errors  
    ACCESS_DENIED: 'Access denied',
    // Validation errors
    VALIDATION_FAILED: 'Validation failed',
    // Not found errors
    RESOURCE_NOT_FOUND: 'Resource not found',
    // Rate limit errors
    RATE_LIMIT_EXCEEDED: 'Rate limit exceeded',
    // Server errors
    INTERNAL_SERVER_ERROR: 'Internal Server error occurred',
    // Network errors
    NETWORK_ERROR: 'Network error occurred',
    REQUEST_TIMEOUT: 'Request timed out',
    REQUEST_ABORTED: 'Request was aborted',
};
/**
 * Error name constants for network error identification
 */
const ErrorNames = {
    ABORT_ERROR: 'AbortError'};

/**
 * Parser for Orchestrator/Task error format
 */
class OrchestratorErrorParser {
    canParse(errorBody) {
        return isOrchestratorError(errorBody);
    }
    parse(errorBody, response) {
        const error = errorBody;
        return {
            message: error.message,
            code: response?.status?.toString(),
            details: {
                errorCode: error.errorCode,
                traceId: error.traceId,
                originalResponse: error
            },
            requestId: error.traceId
        };
    }
}
/**
 * Parser for Entity (Data Fabric) error format
 */
class EntityErrorParser {
    canParse(errorBody) {
        return isEntityError(errorBody);
    }
    parse(errorBody, response) {
        const error = errorBody;
        return {
            message: error.error,
            code: response?.status?.toString(),
            details: {
                error: error.error,
                traceId: error.traceId,
                originalResponse: error
            },
            requestId: error.traceId
        };
    }
}
/**
 * Parser for PIMS error format
 */
class PimsErrorParser {
    canParse(errorBody) {
        return isPimsError(errorBody);
    }
    parse(errorBody, response) {
        const error = errorBody;
        let message = error.title;
        // If there are validation errors, append them to the message for better visibility
        if (error.errors && Object.keys(error.errors).length > 0) {
            const errorMessages = Object.entries(error.errors)
                .map(([field, messages]) => `${field}: ${messages.join(', ')}`)
                .join('; ');
            message += `. Validation errors: ${errorMessages}`;
        }
        return {
            message,
            code: response?.status?.toString(),
            details: {
                type: error.type,
                title: error.title,
                status: error.status,
                errors: error.errors,
                traceId: error.traceId,
                originalResponse: error
            },
            requestId: error.traceId
        };
    }
}
/**
 * Fallback parser for unrecognized formats
 */
class GenericErrorParser {
    canParse(_errorBody) {
        return true; // Always can parse as last resort
    }
    parse(errorBody, response) {
        // For unknown error formats, just pass through the raw error with fallback message
        const message = response?.statusText || 'An error occurred';
        return {
            message,
            code: response?.status?.toString(),
            details: {
                originalResponse: errorBody
            },
        };
    }
}
/**
 * Main error response parser using Chain of Responsibility pattern
 *
 * This parser standardizes error responses from different UiPath services into a
 * consistent format, regardless of the original error structure.
 *
 * Supported formats:
 * 1. Orchestrator/Task: { message, errorCode, traceId }
 * 2. Entity (Data Fabric): { error, traceId }
 * 3. PIMS/Maestro: { type, title, status, errors?, traceId? }
 * 4. Generic: Fallback for any other format
 *
 * @example
 * const parser = new ErrorResponseParser();
 * const errorInfo = await parser.parse(response);
 * // errorInfo will have consistent structure regardless of service
 */
class ErrorResponseParser {
    constructor() {
        this.strategies = [
            new OrchestratorErrorParser(),
            new EntityErrorParser(),
            new PimsErrorParser(),
            new GenericErrorParser() // Must be last
        ];
    }
    /**
     * Parses error response body into standardized format
     * @param response - The HTTP response object
     * @returns Standardized error information
     */
    async parse(response) {
        try {
            const errorBody = await response.json();
            // Find the first strategy that can parse this error format
            const strategy = this.strategies.find(s => s.canParse(errorBody));
            // GenericErrorParser always returns true, so this will never be null
            return strategy.parse(errorBody, response);
        }
        catch {
            // Handle non-JSON responses
            const responseText = await response.text().catch(() => '');
            return {
                message: response.statusText,
                code: response.status.toString(),
                details: {
                    parseError: 'Failed to parse error response as JSON',
                    responseText
                },
                requestId: response.headers.get(HttpHeaders.X_REQUEST_ID) || undefined
            };
        }
    }
}
// Export singleton instance
const errorResponseParser = new ErrorResponseParser();

/**
 * Base error class for all UiPath SDK errors
 * Extends Error for standard error handling compatibility
 */
class UiPathError extends Error {
    constructor(type, params) {
        super(params.message);
        this.name = type;
        this.type = type;
        this.statusCode = params.statusCode;
        this.requestId = params.requestId;
        this.timestamp = new Date();
        // Maintains proper stack trace for where our error was thrown
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, this.constructor);
        }
    }
    /**
     * Returns a clean JSON representation of the error
     */
    toJSON() {
        return {
            type: this.type,
            message: this.message,
            statusCode: this.statusCode,
            requestId: this.requestId,
            timestamp: this.timestamp
        };
    }
    /**
     * Returns detailed debug information including stack trace
     */
    getDebugInfo() {
        return {
            ...this.toJSON(),
            stack: this.stack
        };
    }
}

/**
 * Error thrown when authentication fails (401 errors)
 * Common scenarios:
 * - Invalid credentials
 * - Expired token
 * - Missing authentication
 */
class AuthenticationError extends UiPathError {
    constructor(params = {}) {
        super(ErrorType.AUTHENTICATION, {
            message: params.message || ErrorMessages.AUTHENTICATION_FAILED,
            statusCode: params.statusCode ?? HttpStatus.UNAUTHORIZED,
            requestId: params.requestId
        });
    }
}

/**
 * Error thrown when authorization fails (403 errors)
 * Common scenarios:
 * - Insufficient permissions
 * - Access denied to resource
 * - Invalid scope
 */
class AuthorizationError extends UiPathError {
    constructor(params = {}) {
        super(ErrorType.AUTHORIZATION, {
            message: params.message || ErrorMessages.ACCESS_DENIED,
            statusCode: params.statusCode ?? HttpStatus.FORBIDDEN,
            requestId: params.requestId
        });
    }
}

/**
 * Error thrown when validation fails (400 errors or client-side validation)
 * Common scenarios:
 * - Invalid input parameters
 * - Missing required fields
 * - Invalid data format
 */
class ValidationError extends UiPathError {
    constructor(params = {}) {
        super(ErrorType.VALIDATION, {
            message: params.message || ErrorMessages.VALIDATION_FAILED,
            statusCode: params.statusCode ?? HttpStatus.BAD_REQUEST,
            requestId: params.requestId
        });
    }
}

/**
 * Error thrown when a resource is not found (404 errors)
 * Common scenarios:
 * - Resource doesn't exist
 * - Invalid ID provided
 * - Resource deleted
 */
class NotFoundError extends UiPathError {
    constructor(params = {}) {
        super(ErrorType.NOT_FOUND, {
            message: params.message || ErrorMessages.RESOURCE_NOT_FOUND,
            statusCode: params.statusCode ?? HttpStatus.NOT_FOUND,
            requestId: params.requestId
        });
    }
}

/**
 * Error thrown when rate limit is exceeded (429 errors)
 * Common scenarios:
 * - Too many requests in a time window
 * - API throttling
 */
class RateLimitError extends UiPathError {
    constructor(params = {}) {
        super(ErrorType.RATE_LIMIT, {
            message: params.message || ErrorMessages.RATE_LIMIT_EXCEEDED,
            statusCode: params.statusCode ?? HttpStatus.TOO_MANY_REQUESTS,
            requestId: params.requestId
        });
    }
}

/**
 * Error thrown when server encounters an error (5xx errors)
 * Common scenarios:
 * - Internal server error
 * - Service unavailable
 * - Gateway timeout
 */
class ServerError extends UiPathError {
    constructor(params = {}) {
        super(ErrorType.SERVER, {
            message: params.message || ErrorMessages.INTERNAL_SERVER_ERROR,
            statusCode: params.statusCode ?? HttpStatus.INTERNAL_SERVER_ERROR,
            requestId: params.requestId
        });
    }
    /**
     * Checks if this is a temporary error that might succeed on retry
     */
    get isRetryable() {
        return this.statusCode === HttpStatus.BAD_GATEWAY ||
            this.statusCode === HttpStatus.SERVICE_UNAVAILABLE ||
            this.statusCode === HttpStatus.GATEWAY_TIMEOUT;
    }
}

/**
 * Error thrown when network/connection issues occur
 * Common scenarios:
 * - Connection timeout
 * - DNS resolution failure
 * - Network unreachable
 * - Request aborted
 */
class NetworkError extends UiPathError {
    constructor(params = {}) {
        super(ErrorType.NETWORK, {
            message: params.message || ErrorMessages.NETWORK_ERROR,
            statusCode: params.statusCode, // Network errors typically don't have HTTP status codes
            requestId: params.requestId
        });
    }
}

/**
 * Factory for creating typed errors based on HTTP status codes
 * Follows the Factory pattern for clean error instantiation
 */
class ErrorFactory {
    /**
     * Creates appropriate error instance based on HTTP status code
     */
    static createFromHttpStatus(statusCode, errorInfo) {
        const { message, requestId } = errorInfo;
        // Map status codes to error types
        switch (statusCode) {
            case HttpStatus.BAD_REQUEST:
                return new ValidationError({ message, statusCode, requestId });
            case HttpStatus.UNAUTHORIZED:
                return new AuthenticationError({ message, statusCode, requestId });
            case HttpStatus.FORBIDDEN:
                return new AuthorizationError({ message, statusCode, requestId });
            case HttpStatus.NOT_FOUND:
                return new NotFoundError({ message, statusCode, requestId });
            case HttpStatus.TOO_MANY_REQUESTS:
                return new RateLimitError({ message, statusCode, requestId });
            default:
                // For 5xx errors or any other status code
                if (statusCode >= HttpStatus.INTERNAL_SERVER_ERROR) {
                    return new ServerError({ message, statusCode, requestId });
                }
                // For unknown client errors, treat as validation error
                return new ValidationError({
                    message: `${message} (HTTP ${statusCode})`,
                    statusCode,
                    requestId
                });
        }
    }
    /**
     * Creates a NetworkError from a fetch/network error
     */
    static createNetworkError(error) {
        let message = ErrorMessages.NETWORK_ERROR;
        if (error instanceof Error) {
            if (error.name === ErrorNames.ABORT_ERROR) {
                message = ErrorMessages.REQUEST_ABORTED;
            }
            else if (error.message.includes('timeout')) {
                message = ErrorMessages.REQUEST_TIMEOUT;
            }
            else {
                message = error.message;
            }
        }
        return new NetworkError({ message });
    }
}

const FOLDER_ID = 'X-UIPATH-OrganizationUnitId';
const TRACEPARENT = 'traceparent';
const UIPATH_TRACEPARENT_ID = 'x-uipath-traceparent-id';
/**
 * Content type constants for HTTP requests/responses
 */
const CONTENT_TYPES = {
    JSON: 'application/json',
    XML: 'application/xml',
    OCTET_STREAM: 'application/octet-stream'
};
/**
 * Response type constants for HTTP requests
 */
const RESPONSE_TYPES = {
    JSON: 'json',
    TEXT: 'text',
    BLOB: 'blob',
    ARRAYBUFFER: 'arraybuffer'
};

class ApiClient {
    constructor(config, executionContext, tokenManager, clientConfig = {}) {
        this.config = config;
        this.executionContext = executionContext;
        this.clientConfig = clientConfig;
        this.tokenManager = tokenManager;
    }
    /**
     * Gets a valid authentication token, refreshing if necessary.
     * Used internally for API requests and exposed for services that need manual auth headers.
     *
     * @returns The valid token
     * @throws AuthenticationError if no token available or refresh fails
     */
    async getValidToken() {
        return this.tokenManager.getValidToken();
    }
    async getDefaultHeaders() {
        const token = await this.getValidToken();
        return {
            'Authorization': `Bearer ${token}`,
            'Content-Type': CONTENT_TYPES.JSON,
            ...this.clientConfig.headers
        };
    }
    async request(method, path, options = {}) {
        // Ensure path starts with a forward slash
        const normalizedPath = path.startsWith('/') ? path.substring(1) : path;
        // Construct URL with org and tenant names
        const url = new URL(`${this.config.orgName}/${this.config.tenantName}/${normalizedPath}`, this.config.baseUrl).toString();
        const isFormData = options.body instanceof FormData;
        const defaultHeaders = await this.getDefaultHeaders();
        if (isFormData) {
            delete defaultHeaders['Content-Type'];
        }
        const traceId = crypto.randomUUID().replace(/-/g, '');
        const spanId = crypto.randomUUID().replace(/-/g, '').slice(0, 16);
        const traceparentValue = `00-${traceId}-${spanId}-01`;
        const headers = {
            ...defaultHeaders,
            [TRACEPARENT]: traceparentValue,
            [UIPATH_TRACEPARENT_ID]: traceparentValue,
            ...options.headers
        };
        // Convert params to URLSearchParams
        const searchParams = new URLSearchParams();
        if (options.params) {
            Object.entries(options.params).forEach(([key, value]) => {
                searchParams.append(key, value.toString());
            });
        }
        const fullUrl = searchParams.toString() ? `${url}?${searchParams.toString()}` : url;
        let body = undefined;
        if (options.body) {
            body = isFormData ? options.body : JSON.stringify(options.body);
        }
        try {
            const response = await fetch(fullUrl, {
                method,
                headers,
                body,
                signal: options.signal
            });
            if (!response.ok) {
                const errorInfo = await errorResponseParser.parse(response);
                throw ErrorFactory.createFromHttpStatus(response.status, errorInfo);
            }
            if (response.status === 204) {
                return undefined;
            }
            // Handle blob response type for binary data (e.g., file downloads)
            if (options.responseType === RESPONSE_TYPES.BLOB) {
                const blob = await response.blob();
                return blob;
            }
            // Check if we're expecting XML
            const acceptHeader = headers['Accept'] || headers['accept'];
            if (acceptHeader === CONTENT_TYPES.XML) {
                const text = await response.text();
                return text;
            }
            const text = await response.text();
            if (!text) {
                return undefined;
            }
            try {
                return JSON.parse(text);
            }
            catch (error) {
                if (error instanceof SyntaxError) {
                    throw new ServerError({
                        message: `Server returned non-JSON response (${response.status} ${response.url}): ${error.message}`,
                        statusCode: response.status,
                    });
                }
                throw error;
            }
        }
        catch (error) {
            // If it's already one of our errors, re-throw it
            if (error.type && error.type.includes('Error')) {
                throw error;
            }
            // Otherwise, it's a genuine network/fetch failure
            throw ErrorFactory.createNetworkError(error);
        }
    }
    async get(path, options = {}) {
        return this.request('GET', path, options);
    }
    async post(path, data, options = {}) {
        return this.request('POST', path, { ...options, body: data });
    }
    async put(path, data, options = {}) {
        return this.request('PUT', path, { ...options, body: data });
    }
    async patch(path, data, options = {}) {
        return this.request('PATCH', path, { ...options, body: data });
    }
    async delete(path, options = {}) {
        return this.request('DELETE', path, options);
    }
}

/**
 * Pagination types supported by the SDK
 */
var PaginationType;
(function (PaginationType) {
    PaginationType["OFFSET"] = "offset";
    PaginationType["TOKEN"] = "token";
})(PaginationType || (PaginationType = {}));

/**
 * Collection of utility functions for working with objects
 */
/**
 * Resolves a field value from an object, supporting both direct keys (e.g., '@odata.count')
 * and dot-separated nested paths (e.g., 'pagination.totalCount').
 * Direct key match takes priority over nested traversal.
 */
function resolveNestedField(data, fieldPath) {
    if (!data) {
        return undefined;
    }
    if (fieldPath in data) {
        return data[fieldPath];
    }
    if (!fieldPath.includes('.')) {
        return undefined;
    }
    let value = data;
    for (const part of fieldPath.split('.')) {
        value = value?.[part];
    }
    return value;
}
/**
 * Filters out undefined values from an object
 * @param obj The source object
 * @returns A new object without undefined values
 *
 * @example
 * ```typescript
 * // Object with undefined values
 * const options = {
 *   name: 'test',
 *   count: 5,
 *   prefix: undefined,
 *   suffix: null
 * };
 * const result = filterUndefined(options);
 * // result = { name: 'test', count: 5, suffix: null }
 * ```
 */
function filterUndefined(obj) {
    const result = {};
    for (const [key, value] of Object.entries(obj)) {
        if (value !== undefined) {
            result[key] = value;
        }
    }
    return result;
}

/**
 * Utility functions for platform detection
 */
/**
 * Checks if code is running in a browser environment
 */
const isBrowser = typeof window !== 'undefined' && typeof window.document !== 'undefined';
isBrowser && window.self != window.top && window.location.href.includes('source=ActionCenter');
const _params = isBrowser ? new URLSearchParams(window.location.search) : null;
/**
 * True when the coded app has been loaded inside a host frame that explicitly
 * opted into token delegation by adding `?host=embed` to the iframe src URL.
 */
const isHostEmbedded = isBrowser && window.self !== window.top && _params?.get('host') === 'embed';
/**
 * The validated parent origin, read from the `?basedomain=` query param set
 * by the embedding host in the iframe src URL.
 * Mirrors the same mechanism used by ActionCenterTokenManager.
 * Non-null only when `?host=embed` is present and `?basedomain=` is a valid URL.
 */
(() => {
    if (!isHostEmbedded)
        return null;
    const basedomain = _params?.get('basedomain');
    if (!basedomain)
        return null;
    try {
        return new URL(basedomain).origin;
    }
    catch {
        console.warn('embeddingOrigin: basedomain query param is not a valid URL', basedomain);
        return null;
    }
})();

/**
 * Base64 encoding/decoding
 */
/**
 * Encodes a string to base64
 * @param str - The string to encode
 * @returns Base64 encoded string
 */
function encodeBase64(str) {
    // TextEncoder for UTF-8 encoding (works in both browser and Node.js)
    const encoder = new TextEncoder();
    const data = encoder.encode(str);
    // Convert Uint8Array to base64
    if (isBrowser) {
        // Browser environment
        // Convert Uint8Array to binary string then to base64
        const binaryString = Array.from(data, byte => String.fromCharCode(byte)).join('');
        return btoa(binaryString);
    }
    else {
        // Node.js environment
        return Buffer.from(data).toString('base64');
    }
}
/**
 * Decodes a base64 string
 * @param base64 - The base64 string to decode
 * @returns Decoded string
 */
function decodeBase64(base64) {
    let bytes;
    if (isBrowser) {
        // Browser environment
        const binaryString = atob(base64);
        bytes = new Uint8Array(binaryString.length);
        for (let i = 0; i < binaryString.length; i++) {
            bytes[i] = binaryString.charCodeAt(i);
        }
    }
    else {
        // Node.js environment
        bytes = new Uint8Array(Buffer.from(base64, 'base64'));
    }
    // TextDecoder for UTF-8 decoding (works in both browser and Node.js)
    const decoder = new TextDecoder();
    return decoder.decode(bytes);
}

/**
 * PaginationManager handles the conversion between uniform cursor-based pagination
 * and the specific pagination type for each service
 */
class PaginationManager {
    /**
     * Create a pagination cursor for subsequent page requests
     */
    static createCursor({ pageInfo, type }) {
        if (!pageInfo.hasMore) {
            return undefined;
        }
        const cursorData = {
            type,
            pageSize: pageInfo.pageSize,
        };
        switch (type) {
            case PaginationType.OFFSET:
                if (pageInfo.currentPage) {
                    cursorData.pageNumber = pageInfo.currentPage + 1;
                }
                break;
            case PaginationType.TOKEN:
                if (pageInfo.continuationToken) {
                    cursorData.continuationToken = pageInfo.continuationToken;
                }
                else {
                    return undefined; // No continuation token, can't continue
                }
                break;
        }
        return {
            value: encodeBase64(JSON.stringify(cursorData))
        };
    }
    /**
     * Create a paginated response with navigation cursors
     */
    static createPaginatedResponse({ pageInfo, type }, items) {
        const nextCursor = PaginationManager.createCursor({ pageInfo, type });
        // Create previous page cursor if applicable
        let previousCursor = undefined;
        if (pageInfo.currentPage && pageInfo.currentPage > 1) {
            const prevCursorData = {
                type,
                pageNumber: pageInfo.currentPage - 1,
                pageSize: pageInfo.pageSize,
            };
            previousCursor = {
                value: encodeBase64(JSON.stringify(prevCursorData))
            };
        }
        // Calculate total pages if we have totalCount and pageSize
        let totalPages = undefined;
        if (pageInfo.totalCount !== undefined && pageInfo.pageSize) {
            totalPages = Math.ceil(pageInfo.totalCount / pageInfo.pageSize);
        }
        // Determine if this pagination type supports page jumping
        const supportsPageJump = type === PaginationType.OFFSET;
        // Create the result object with all fields, then filter out undefined values
        const result = filterUndefined({
            items,
            totalCount: pageInfo.totalCount,
            hasNextPage: pageInfo.hasMore,
            nextCursor: nextCursor,
            previousCursor: previousCursor,
            currentPage: pageInfo.currentPage,
            totalPages,
            supportsPageJump
        });
        return result;
    }
}

/**
 * Creates headers object from key-value pairs
 * @param headersObj - Object containing header key-value pairs
 * @returns Headers object with all values converted to strings
 *
 * @example
 * ```typescript
 * // Single header
 * const headers = createHeaders({ 'X-UIPATH-FolderKey': '1234567890' });
 *
 * // Multiple headers
 * const headers = createHeaders({
 *   'X-UIPATH-FolderKey': '1234567890',
 *   'X-UIPATH-OrganizationUnitId': 123,
 *   'Accept': 'application/json'
 * });
 *
 * // Using constants
 * import { FOLDER_KEY, FOLDER_ID } from '../constants/headers';
 * const headers = createHeaders({
 *   [FOLDER_KEY]: 'abc-123',
 *   [FOLDER_ID]: 456
 * });
 *
 * // Empty headers
 * const headers = createHeaders();
 * ```
 */
function createHeaders(headersObj) {
    const headers = {};
    for (const [key, value] of Object.entries(headersObj)) {
        if (value !== undefined && value !== null) {
            headers[key] = value.toString();
        }
    }
    return headers;
}

/**
 * Common constants used across the SDK
 */
/**
 * Prefix used for OData query parameters
 */
const ODATA_PREFIX = '$';
/**
 * HTTP methods
 */
const HTTP_METHODS = {
    GET: 'GET',
    POST: 'POST'};
/**
 * Agents pagination constants — items and total count are nested under the response envelope
 */
const AGENTS_PAGINATION = {
    /** Dotted path to the total count in the agents response envelope */
    TOTAL_COUNT_FIELD: 'pagination.totalCount'
};
/**
 * Agents OFFSET pagination parameter names (page-number style, 0-based, no skip conversion)
 */
const AGENTS_OFFSET_PARAMS = {
    /** Page size parameter name */
    PAGE_SIZE_PARAM: 'pageSize',
    /** Page number parameter name (sent directly, 0-based) */
    OFFSET_PARAM: 'pageNumber',
    /** No count param needed */
    COUNT_PARAM: undefined
};
/**
 * Traceview spans pagination constants — items sit directly under `data`,
 * total count under the same envelope location as the agents list. Request
 * params reuse {@link AGENTS_OFFSET_PARAMS} (pageSize + 0-based pageNumber).
 */
const TRACEVIEW_SPANS_PAGINATION = {
    /** Field name for the spans array in the response envelope */
    ITEMS_FIELD: 'data',
    /** Total count path — same envelope location as the agents list. */
    TOTAL_COUNT_FIELD: AGENTS_PAGINATION.TOTAL_COUNT_FIELD
};
/**
 * OData OFFSET pagination parameter names (ODATA-style)
 */
const ODATA_OFFSET_PARAMS = {
    /** OData page size parameter name */
    PAGE_SIZE_PARAM: '$top',
    /** OData offset parameter name */
    OFFSET_PARAM: '$skip',
    /** OData count parameter name */
    COUNT_PARAM: '$count'
};
/**
 * Bucket TOKEN pagination parameter names
 */
const BUCKET_TOKEN_PARAMS = {
    /** Bucket page size parameter name */
    PAGE_SIZE_PARAM: 'takeHint',
    /** Bucket token parameter name */
    TOKEN_PARAM: 'continuationToken'
};

/**
 * Converts a UTC timestamp string (e.g., "5/8/2026 11:20:17 AM") to ISO 8601 UTC format.
 * Returns the original value if parsing fails.
 */
/**
 * Converts a string from PascalCase to camelCase
 * @param str The PascalCase string to convert
 * @returns The camelCase version of the string
 *
 * @example
 * ```typescript
 * pascalToCamelCase('HelloWorld'); // 'helloWorld'
 * pascalToCamelCase('TaskAssignmentCriteria'); // 'taskAssignmentCriteria'
 * ```
 */
function pascalToCamelCase(str) {
    if (!str)
        return str;
    return str.charAt(0).toLowerCase() + str.slice(1);
}
/**
 * Generic function to transform object keys using a provided case conversion function
 * @param data The object to transform
 * @param convertCase The function to convert each key
 * @returns A new object with transformed keys
 */
function transformCaseKeys(data, convertCase) {
    // Handle array of objects
    if (Array.isArray(data)) {
        return data.map(item => {
            // If the array element is a primitive (string, number, etc.), return it as is
            if (item === null || typeof item !== 'object' || typeof item === 'string') {
                return item;
            }
            // Only recursively transform if it's actually an object
            return transformCaseKeys(item, convertCase);
        });
    }
    const result = {};
    for (const [key, value] of Object.entries(data)) {
        const transformedKey = convertCase(key);
        // Recursively transform nested objects and arrays
        if (value !== null && typeof value === 'object') {
            result[transformedKey] = transformCaseKeys(value, convertCase);
        }
        else {
            result[transformedKey] = value;
        }
    }
    return result;
}
/**
 * Transforms an object's keys from PascalCase to camelCase
 * @param data The object with PascalCase keys
 * @returns A new object with all keys converted to camelCase
 *
 * @example
 * ```typescript
 * // Simple object
 * pascalToCamelCaseKeys({ Id: "123", TaskName: "Invoice" });
 * // Result: { id: "123", taskName: "Invoice" }
 *
 * // Nested object
 * pascalToCamelCaseKeys({
 *   TaskId: "456",
 *   TaskDetails: { AssignedUser: "John", Priority: "High" }
 * });
 * // Result: {
 * //   taskId: "456",
 * //   taskDetails: { assignedUser: "John", priority: "High" }
 * // }
 *
 * // Array of objects
 * pascalToCamelCaseKeys([
 *   { Id: "1", IsComplete: false },
 *   { Id: "2", IsComplete: true }
 * ]);
 * // Result: [
 * //   { id: "1", isComplete: false },
 * //   { id: "2", isComplete: true }
 * // ]
 * ```
 */
function pascalToCamelCaseKeys(data) {
    return transformCaseKeys(data, pascalToCamelCase);
}
/**
 * Adds a prefix to specified keys in an object, returning a new object.
 * Only the provided keys are prefixed; all others are left unchanged.
 *
 * @param obj The source object
 * @param prefix The prefix to add (e.g., '$')
 * @param keys The keys to prefix (e.g., ['expand', 'filter'])
 * @returns A new object with specified keys prefixed
 *
 * @example
 * addPrefixToKeys({ expand: 'a', foo: 1 }, '$', ['expand']) // { $expand: 'a', foo: 1 }
 */
function addPrefixToKeys(obj, prefix, keys) {
    const result = {};
    for (const [key, value] of Object.entries(obj)) {
        if (keys.includes(key)) {
            result[`${prefix}${key}`] = value;
        }
        else {
            result[key] = value;
        }
    }
    return result;
}

/**
 * Constants used throughout the pagination system
 */
/** Maximum number of items that can be requested in a single page */
const MAX_PAGE_SIZE = 1000;
/** Default page size when jumpToPage is used without specifying pageSize */
const DEFAULT_PAGE_SIZE = 50;
/** Default field name for items in a paginated response */
const DEFAULT_ITEMS_FIELD = 'value';
/** Default field name for total count in a paginated response */
const DEFAULT_TOTAL_COUNT_FIELD = '@odata.count';
/**
 * Limits the page size to the maximum allowed value
 * @param pageSize - Requested page size
 * @returns Limited page size value
 */
function getLimitedPageSize(pageSize) {
    if (pageSize === undefined || pageSize === null) {
        return DEFAULT_PAGE_SIZE;
    }
    return Math.max(1, Math.min(pageSize, MAX_PAGE_SIZE));
}

/**
 * Helper functions for pagination that can be used across services
 */
class PaginationHelpers {
    /**
     * Checks if any pagination parameters are provided
     *
     * @param options - The options object to check
     * @returns True if any pagination parameter is defined, false otherwise
     */
    static hasPaginationParameters(options = {}) {
        const { cursor, pageSize, jumpToPage } = options;
        return cursor !== undefined || pageSize !== undefined || jumpToPage !== undefined;
    }
    /**
     * Parse a pagination cursor string into cursor data
     */
    static parseCursor(cursorString) {
        try {
            const cursorData = JSON.parse(decodeBase64(cursorString));
            return cursorData;
        }
        catch {
            throw new Error('Invalid pagination cursor');
        }
    }
    /**
     * Validates cursor format and structure
     *
     * @param paginationOptions - The pagination options containing the cursor
     * @param paginationType - Optional pagination type to validate against
     */
    static validateCursor(paginationOptions, paginationType) {
        if (paginationOptions.cursor !== undefined) {
            if (!paginationOptions.cursor || typeof paginationOptions.cursor.value !== 'string' || !paginationOptions.cursor.value) {
                throw new Error('cursor must contain a valid cursor string');
            }
            try {
                // Try to parse the cursor to validate it
                const cursorData = PaginationHelpers.parseCursor(paginationOptions.cursor.value);
                // If type is provided, validate cursor contains expected type information
                if (paginationType) {
                    if (!cursorData.type) {
                        throw new Error('Invalid cursor: missing pagination type');
                    }
                    // Check pagination type compatibility
                    if (cursorData.type !== paginationType) {
                        throw new Error(`Pagination type mismatch: cursor is for ${cursorData.type} but service uses ${paginationType}`);
                    }
                }
            }
            catch (error) {
                if (error instanceof Error) {
                    // If it's already our error with specific message, pass it through
                    if (error.message.startsWith('Invalid cursor') ||
                        error.message.startsWith('Pagination type mismatch')) {
                        throw error;
                    }
                }
                throw new Error('Invalid pagination cursor format');
            }
        }
    }
    /**
     * Comprehensive validation for pagination options
     *
     * @param options - The pagination options to validate
     * @param paginationType - The pagination type these options will be used with
     * @returns Processed pagination parameters ready for use
     */
    static validatePaginationOptions(options, paginationType) {
        // Validate pageSize
        if (options.pageSize !== undefined && options.pageSize <= 0) {
            throw new Error('pageSize must be a positive number');
        }
        // Validate jumpToPage
        if (options.jumpToPage !== undefined && options.jumpToPage <= 0) {
            throw new Error('jumpToPage must be a positive number');
        }
        // Validate cursor
        PaginationHelpers.validateCursor(options, paginationType);
        // Validate service compatibility
        if (options.jumpToPage !== undefined && paginationType === PaginationType.TOKEN) {
            throw new Error('jumpToPage is not supported for token-based pagination. Use cursor-based navigation instead.');
        }
        // Get processed parameters
        return PaginationHelpers.getRequestParameters(options, paginationType);
    }
    /**
     * Convert a unified pagination options to service-specific parameters
     */
    static getRequestParameters(options, paginationType) {
        // Handle jumpToPage
        if (options.jumpToPage !== undefined) {
            const jumpToPageOptions = {
                pageSize: options.pageSize,
                pageNumber: options.jumpToPage
            };
            return filterUndefined(jumpToPageOptions);
        }
        // If no cursor is provided, it's a first page request
        if (!options.cursor) {
            const firstPageOptions = {
                pageSize: options.pageSize,
                // Only set pageNumber for OFFSET pagination
                pageNumber: paginationType === PaginationType.OFFSET ? 1 : undefined
            };
            return filterUndefined(firstPageOptions);
        }
        // Parse the cursor
        try {
            const cursorData = PaginationHelpers.parseCursor(options.cursor.value);
            const cursorBasedOptions = {
                pageSize: cursorData.pageSize || options.pageSize,
                pageNumber: cursorData.pageNumber,
                continuationToken: cursorData.continuationToken,
                type: cursorData.type,
            };
            return filterUndefined(cursorBasedOptions);
        }
        catch {
            throw new Error('Invalid pagination cursor');
        }
    }
    /**
     * Helper method for paginated resource retrieval
     *
     * @param params - Parameters for pagination
     * @returns Promise resolving to a paginated result
     */
    static async getAllPaginated(params) {
        const { serviceAccess, getEndpoint, folderId, headers: providedHeaders, paginationParams, additionalParams, queryParams, transformFn, method = HTTP_METHODS.GET, options = {} } = params;
        const endpoint = getEndpoint(folderId);
        const headers = providedHeaders ?? (folderId ? createHeaders({ [FOLDER_ID]: folderId }) : {});
        // On POST, the caller's options go in the body; queryParams stays in the URL.
        // On GET, everything is URL — queryParams merges with additionalParams.
        const isPost = method === HTTP_METHODS.POST;
        const requestSpec = isPost
            ? { body: additionalParams, params: queryParams }
            : { params: { ...additionalParams, ...queryParams } };
        const paginatedResponse = await serviceAccess.requestWithPagination(method, endpoint, paginationParams, {
            headers,
            ...requestSpec,
            pagination: {
                paginationType: options.paginationType || PaginationType.OFFSET,
                itemsField: options.itemsField || DEFAULT_ITEMS_FIELD,
                totalCountField: options.totalCountField || DEFAULT_TOTAL_COUNT_FIELD,
                continuationTokenField: options.continuationTokenField,
                paginationParams: options.paginationParams
            }
        });
        // Parse items - automatically handle JSON string responses
        const rawItems = paginatedResponse.items;
        const parsedItems = typeof rawItems === 'string' ? JSON.parse(rawItems) : (rawItems || []);
        const transformedItems = transformFn ? parsedItems.map(transformFn) : parsedItems;
        return {
            ...paginatedResponse,
            items: transformedItems
        };
    }
    /**
     * Helper method for non-paginated resource retrieval
     *
     * @param params - Parameters for non-paginated resource retrieval
     * @returns Promise resolving to an object with data and totalCount
     */
    static async getAllNonPaginated(params) {
        const { serviceAccess, getAllEndpoint, getByFolderEndpoint, folderId, headers: providedHeaders, additionalParams, queryParams, transformFn, method = HTTP_METHODS.GET, options = {} } = params;
        // Set default field names
        const itemsField = options.itemsField || DEFAULT_ITEMS_FIELD;
        const totalCountField = options.totalCountField || DEFAULT_TOTAL_COUNT_FIELD;
        // Determine endpoint and headers based on folderId
        const endpoint = folderId ? getByFolderEndpoint : getAllEndpoint;
        const headers = providedHeaders ?? (folderId ? createHeaders({ [FOLDER_ID]: folderId }) : {});
        // Make the API call based on method
        let response;
        if (method === HTTP_METHODS.POST) {
            response = await serviceAccess.post(endpoint, additionalParams, { headers, params: queryParams });
        }
        else {
            response = await serviceAccess.get(endpoint, {
                params: { ...additionalParams, ...queryParams },
                headers
            });
        }
        // Extract and transform items from response
        // Handle both plain array responses and envelope responses ({ value: [...], totalRecordCount: N }).
        // itemsField may be a dotted path (e.g. 'data.agents') for nested envelopes.
        const rawItems = Array.isArray(response.data) ? response.data : resolveNestedField(response.data, itemsField);
        const rawTotalCount = Array.isArray(response.data) ? undefined : resolveNestedField(response.data, totalCountField);
        const totalCount = typeof rawTotalCount === 'number' ? rawTotalCount : undefined;
        // Parse items - automatically handle JSON string responses
        const parsedItems = typeof rawItems === 'string' ? JSON.parse(rawItems) : (rawItems || []);
        const items = transformFn ? parsedItems.map(transformFn) : parsedItems;
        return {
            items,
            totalCount
        };
    }
    /**
     * Centralized getAll implementation that handles both paginated and non-paginated requests
     *
     * @param config - Configuration for the getAll operation
     * @param options - Request options including pagination parameters
     * @returns Promise resolving to either paginated or non-paginated response based on options
     */
    static async getAll(config, options) {
        const optionsWithDefaults = options || {};
        const { folderId, pageSize, cursor, jumpToPage, ...restOptions } = optionsWithDefaults;
        // Determine if pagination is requested
        const isPaginationRequested = PaginationHelpers.hasPaginationParameters(options || {});
        // Process parameters (custom processing if provided, otherwise default)
        let processedOptions = restOptions;
        if (config.processParametersFn) {
            processedOptions = config.processParametersFn(restOptions, folderId);
        }
        // Apply ODATA prefix to keys (excluding specified keys)
        const excludeKeys = config.excludeFromPrefix || [];
        const keysToPrefix = Object.keys(processedOptions).filter(k => !excludeKeys.includes(k));
        const prefixedOptions = addPrefixToKeys(processedOptions, ODATA_PREFIX, keysToPrefix);
        // Default pagination options
        const paginationOptions = {
            paginationType: PaginationType.OFFSET,
            itemsField: DEFAULT_ITEMS_FIELD,
            totalCountField: DEFAULT_TOTAL_COUNT_FIELD,
            ...config.pagination
        };
        // Paginated flow
        if (isPaginationRequested) {
            return PaginationHelpers.getAllPaginated({
                serviceAccess: config.serviceAccess,
                getEndpoint: config.getEndpoint,
                folderId,
                headers: config.headers,
                paginationParams: cursor ? { cursor, pageSize } : jumpToPage !== undefined ? { jumpToPage, pageSize } : { pageSize },
                additionalParams: prefixedOptions,
                queryParams: config.queryParams,
                transformFn: config.transformFn,
                method: config.method,
                options: {
                    ...paginationOptions,
                    paginationParams: config.pagination?.paginationParams
                }
            }); // Type assertion needed due to conditional return
        }
        // Non-paginated flow
        const byFolderEndpoint = config.getByFolderEndpoint || config.getEndpoint(folderId);
        return PaginationHelpers.getAllNonPaginated({
            serviceAccess: config.serviceAccess,
            getAllEndpoint: config.getEndpoint(),
            getByFolderEndpoint: byFolderEndpoint,
            folderId,
            headers: config.headers,
            additionalParams: prefixedOptions,
            queryParams: config.queryParams,
            transformFn: config.transformFn,
            method: config.method,
            options: {
                itemsField: paginationOptions.itemsField,
                totalCountField: paginationOptions.totalCountField
            }
        });
    }
}

/**
 * SDK Internals Registry - Internal registry for SDK instances
 *
 * This class is NOT exported in the public API.
 * It provides a secure way to share SDK internals between
 * the UiPath class and service classes without exposing them publicly.
 *
 * @internal
 */
// Global symbol key to ensure WeakMap is shared across module instances
// This prevents issues when core and service modules are bundled separately
const REGISTRY_KEY = Symbol.for('@uipath/sdk-internals-registry');
// Get or create the global WeakMap store
const getGlobalStore = () => {
    const globalObj = globalThis;
    if (!globalObj[REGISTRY_KEY]) {
        globalObj[REGISTRY_KEY] = new WeakMap();
    }
    return globalObj[REGISTRY_KEY];
};
/**
 * Internal registry for SDK private components.
 * Uses WeakMap to prevent memory leaks - entries are automatically
 * garbage collected when the SDK instance is no longer referenced.
 *
 * Uses a global singleton pattern to ensure the same WeakMap is shared
 * across separately bundled modules (core, entities, tasks, etc.).
 *
 * @internal - Not exported in public API
 */
class SDKInternalsRegistry {
    // Use global store to ensure sharing across module bundles
    static get store() {
        return getGlobalStore();
    }
    /**
     * Register SDK instance internals
     * Called by UiPath constructor
     */
    static set(instance, internals) {
        this.store.set(instance, internals);
    }
    /**
     * Retrieve SDK instance internals
     * Called by BaseService constructor
     */
    static get(instance) {
        const internals = this.store.get(instance);
        if (!internals) {
            throw new Error('Invalid SDK instance. Make sure to pass a valid UiPath instance to the service constructor.');
        }
        return internals;
    }
}

var _BaseService_apiClient;
/**
 * Base class for all UiPath SDK services.
 *
 * Provides common functionality for authentication, configuration, and API communication.
 * All service classes extend this base to inherit dependency injection and HTTP client access.
 *
 * This class implements the dependency injection pattern where services receive a configured
 * UiPath instance. The ApiClient is created internally and handles all HTTP operations
 * including authentication token management.
 *
 * @remarks
 * Service classes should extend this base and call `super(uiPath)` in their constructor.
 * Protected HTTP methods (get, post, put, patch, delete) are available to all subclasses.
 *
 */
class BaseService {
    /**
     * Creates a base service instance with dependency injection.
     *
     * Extracts configuration, execution context, and token manager from the UiPath instance
     * to initialize an authenticated API client. The ApiClient handles all HTTP operations
     * and token management internally.
     *
     * @param instance - UiPath SDK instance providing authentication and configuration.
     *                    Services receive this via dependency injection in the modular pattern.
     * @param headers - Optional default headers to include in every request (e.g. `x-uipath-external-user-id` for
     *                  CAS external-app auth)
     *
     * @example
     * ```typescript
     * // Services automatically call this via super()
     * export class EntityService extends BaseService {
     *   constructor(instance: IUiPath) {
     *     super(instance); // Initializes the internal ApiClient
     *   }
     * }
     *
     * // Usage in modular pattern
     * import { UiPath } from '@uipath/uipath-typescript/core';
     * import { Entities } from '@uipath/uipath-typescript/entities';
     *
     * const sdk = new UiPath(config);
     * await sdk.initialize();
     * const entities = new Entities(sdk);
     * ```
     */
    constructor(instance, headers) {
        // Private field - not visible via Object.keys() or any reflection
        _BaseService_apiClient.set(this, void 0);
        const { config, context, tokenManager, folderKey } = SDKInternalsRegistry.get(instance);
        __classPrivateFieldSet(this, _BaseService_apiClient, new ApiClient(config, context, tokenManager, headers ? { headers } : {}), "f");
        this.config = { folderKey };
    }
    /**
     * Gets a valid authentication token, refreshing if necessary.
     * Use this when you need to manually add Authorization headers (e.g., direct uploads).
     *
     * @returns Promise resolving to a valid access token string
     * @throws AuthenticationError if no token is available or refresh fails
     */
    async getValidAuthToken() {
        return __classPrivateFieldGet(this, _BaseService_apiClient, "f").getValidToken();
    }
    /**
     * Creates a service accessor for pagination helpers
     * This allows pagination helpers to access protected methods without making them public
     */
    createPaginationServiceAccess() {
        return {
            get: (path, options) => this.get(path, options || {}),
            post: (path, body, options) => this.post(path, body, options || {}),
            requestWithPagination: (method, path, paginationOptions, options) => this.requestWithPagination(method, path, paginationOptions, options)
        };
    }
    async request(method, path, options = {}) {
        switch (method.toUpperCase()) {
            case 'GET':
                return this.get(path, options);
            case 'POST':
                return this.post(path, options.body, options);
            case 'PUT':
                return this.put(path, options.body, options);
            case 'PATCH':
                return this.patch(path, options.body, options);
            case 'DELETE':
                return this.delete(path, options);
            default:
                throw new Error(`Unsupported HTTP method: ${method}`);
        }
    }
    async requestWithSpec(spec) {
        if (!spec.method || !spec.url) {
            throw new Error('Request spec must include method and url');
        }
        return this.request(spec.method, spec.url, spec);
    }
    async get(path, options = {}) {
        const response = await __classPrivateFieldGet(this, _BaseService_apiClient, "f").get(path, options);
        return { data: response };
    }
    async post(path, data, options = {}) {
        const response = await __classPrivateFieldGet(this, _BaseService_apiClient, "f").post(path, data, options);
        return { data: response };
    }
    async put(path, data, options = {}) {
        const response = await __classPrivateFieldGet(this, _BaseService_apiClient, "f").put(path, data, options);
        return { data: response };
    }
    async patch(path, data, options = {}) {
        const response = await __classPrivateFieldGet(this, _BaseService_apiClient, "f").patch(path, data, options);
        return { data: response };
    }
    async delete(path, options = {}) {
        const response = await __classPrivateFieldGet(this, _BaseService_apiClient, "f").delete(path, options);
        return { data: response };
    }
    /**
     * Execute a request with cursor-based pagination
     */
    async requestWithPagination(method, path, paginationOptions, options) {
        const paginationType = options.pagination.paginationType;
        // Validate and prepare pagination parameters
        const params = this.validateAndPreparePaginationParams(paginationType, paginationOptions);
        // Prepare request parameters based on pagination type
        const requestParams = this.preparePaginationRequestParams(paginationType, params, options.pagination);
        // Route pagination state to wherever the API expects it (body for POST, URL for GET).
        // Caller-supplied options.body / options.params are respected as-is — the api-client
        // already handles params (URL) and body (request body) independently for every method.
        if (method.toUpperCase() === 'POST') {
            const existingBody = (options.body && typeof options.body === 'object') ? options.body : {};
            options.body = {
                ...existingBody,
                ...requestParams
            };
        }
        else {
            options.params = {
                ...options.params,
                ...requestParams
            };
        }
        // Make the request
        const response = await this.request(method, path, options);
        // Extract data from the response and create page result
        return this.createPaginatedResponseFromResponse(response, params, paginationType, {
            itemsField: options.pagination.itemsField,
            totalCountField: options.pagination.totalCountField,
            continuationTokenField: options.pagination.continuationTokenField
        });
    }
    /**
     * Validates and prepares pagination parameters from options
     */
    validateAndPreparePaginationParams(paginationType, paginationOptions) {
        return PaginationHelpers.validatePaginationOptions(paginationOptions, paginationType);
    }
    /**
     * Prepares request parameters for pagination based on pagination type
     */
    preparePaginationRequestParams(paginationType, params, paginationConfig) {
        const requestParams = {};
        let limitedPageSize;
        const paginationParams = paginationConfig?.paginationParams;
        switch (paginationType) {
            case PaginationType.OFFSET:
                limitedPageSize = getLimitedPageSize(params.pageSize);
                const pageSizeParam = paginationParams?.pageSizeParam || ODATA_OFFSET_PARAMS.PAGE_SIZE_PARAM;
                const offsetParam = paginationParams?.offsetParam || ODATA_OFFSET_PARAMS.OFFSET_PARAM;
                const countParam = paginationParams?.countParam || ODATA_OFFSET_PARAMS.COUNT_PARAM;
                // When true (default), converts pageNumber to a skip/offset value (e.g., page 3 with pageSize 10 → skip 20).
                // When false, passes pageNumber directly as the offset param — used by APIs that accept a page number instead of a record offset.
                const convertToSkip = paginationParams?.convertToSkip ?? true;
                // When true, sends pageNumber - 1 (for 0-based APIs). Default false (1-based).
                const zeroBased = paginationParams?.zeroBased ?? false;
                requestParams[pageSizeParam] = limitedPageSize;
                if (convertToSkip) {
                    if (params.pageNumber && params.pageNumber > 1) {
                        requestParams[offsetParam] = (params.pageNumber - 1) * limitedPageSize;
                    }
                }
                else {
                    const sdkPageNumber = params.pageNumber || 1;
                    requestParams[offsetParam] = zeroBased ? sdkPageNumber - 1 : sdkPageNumber;
                }
                {
                    requestParams[countParam] = true;
                }
                break;
            case PaginationType.TOKEN:
                const tokenPageSizeParam = paginationParams?.pageSizeParam || BUCKET_TOKEN_PARAMS.PAGE_SIZE_PARAM;
                const tokenParam = paginationParams?.tokenParam || BUCKET_TOKEN_PARAMS.TOKEN_PARAM;
                if (params.pageSize) {
                    requestParams[tokenPageSizeParam] = getLimitedPageSize(params.pageSize);
                }
                if (params.continuationToken) {
                    requestParams[tokenParam] = params.continuationToken;
                }
                break;
        }
        return requestParams;
    }
    /**
     * Creates a paginated response from API response
     */
    createPaginatedResponseFromResponse(response, params, paginationType, fields) {
        // Extract fields from response
        const itemsField = fields.itemsField ||
            (paginationType === PaginationType.TOKEN ? 'items' : 'value');
        const totalCountField = fields.totalCountField || 'totalRecordCount';
        const continuationTokenField = fields.continuationTokenField || 'continuationToken';
        // Extract items and metadata
        // Handle both plain array responses and envelope responses ({ value: [...], totalRecordCount: N }).
        // itemsField may be a dotted path (e.g. 'data.agents') for nested envelopes.
        const items = Array.isArray(response.data) ? response.data : (resolveNestedField(response.data, itemsField) || []);
        const rawTotalCount = Array.isArray(response.data) ? undefined : resolveNestedField(response.data, totalCountField);
        const totalCount = typeof rawTotalCount === 'number' ? rawTotalCount : undefined;
        const continuationToken = response.data[continuationTokenField];
        // Determine if there are more pages
        const hasMore = this.determineHasMorePages(paginationType, {
            totalCount,
            pageSize: params.pageSize,
            currentPage: params.pageNumber || 1,
            itemsCount: items.length,
            continuationToken
        });
        // Create and return the page result
        return PaginationManager.createPaginatedResponse({
            pageInfo: {
                hasMore,
                totalCount,
                currentPage: params.pageNumber,
                pageSize: params.pageSize,
                continuationToken
            },
            type: paginationType,
        }, items);
    }
    /**
     * Determines if there are more pages based on pagination type and metadata
     */
    determineHasMorePages(paginationType, info) {
        switch (paginationType) {
            case PaginationType.OFFSET:
                const effectivePageSize = info.pageSize ?? DEFAULT_PAGE_SIZE;
                // If totalCount is available, use it for precise calculation
                if (info.totalCount !== undefined) {
                    return (info.currentPage * effectivePageSize) < info.totalCount;
                }
                // Fallback when totalCount is not available
                // NOTE: This code path should rarely be executed as the APIs typically return totalCount
                return info.itemsCount === effectivePageSize;
            case PaginationType.TOKEN:
                return !!info.continuationToken;
            default:
                return false;
        }
    }
}
_BaseService_apiClient = new WeakMap();

/** Status of a span: whether it completed successfully, with an error, or was not set. */
var SpanStatus;
(function (SpanStatus) {
    SpanStatus["Unset"] = "Unset";
    SpanStatus["Ok"] = "Ok";
    SpanStatus["Error"] = "Error";
    /** Span is still in progress. */
    SpanStatus["Running"] = "Running";
    /** Span data is hidden from the caller due to tenant/folder permission rules. */
    SpanStatus["Restricted"] = "Restricted";
    /** Span was cancelled before completion. */
    SpanStatus["Cancelled"] = "Cancelled";
})(SpanStatus || (SpanStatus = {}));
/** Platform source that produced the span. */
var SpanSource;
(function (SpanSource) {
    SpanSource["Testing"] = "Testing";
    SpanSource["Agents"] = "Agents";
    SpanSource["ProcessOrchestration"] = "ProcessOrchestration";
    SpanSource["ApiWorkflows"] = "ApiWorkflows";
    SpanSource["Robots"] = "Robots";
    SpanSource["ConversationalAgentsService"] = "ConversationalAgentsService";
    SpanSource["IntegrationServiceTrigger"] = "IntegrationServiceTrigger";
    SpanSource["Playground"] = "Playground";
    SpanSource["Governance"] = "Governance";
    /** Intelligent Experience Platform — unstructured and complex document processing source. */
    SpanSource["IXPUnstructuredAndComplexDocuments"] = "IXPUnstructuredAndComplexDocuments";
    /** Agents authored in code (as opposed to visual/no-code designers). */
    SpanSource["CodedAgents"] = "CodedAgents";
    /** Intelligent Experience Platform — communications mining source. */
    SpanSource["IXPCommunicationsMining"] = "IXPCommunicationsMining";
    /** UiPath Context Grounding — span produced by the Enterprise Context Service for RAG/knowledge-base operations. */
    SpanSource["EnterpriseContextService"] = "EnterpriseContextService";
    /** Model Context Protocol — span produced by an MCP server integration. */
    SpanSource["MCP"] = "MCP";
    /** Agent-to-Agent — span produced by an A2A protocol call between agents. */
    SpanSource["A2A"] = "A2A";
    /** Serverless — span produced by a serverless function execution. */
    SpanSource["Serverless"] = "Serverless";
})(SpanSource || (SpanSource = {}));
/** Minimum severity level of events captured in the span. */
var SpanVerbosityLevel;
(function (SpanVerbosityLevel) {
    SpanVerbosityLevel["Verbose"] = "Verbose";
    SpanVerbosityLevel["Trace"] = "Trace";
    SpanVerbosityLevel["Information"] = "Information";
    SpanVerbosityLevel["Warning"] = "Warning";
    SpanVerbosityLevel["Error"] = "Error";
    SpanVerbosityLevel["Critical"] = "Critical";
    SpanVerbosityLevel["Off"] = "Off";
})(SpanVerbosityLevel || (SpanVerbosityLevel = {}));
/** Whether the span was produced during a debug or production runtime. */
var SpanExecutionType;
(function (SpanExecutionType) {
    SpanExecutionType["Debug"] = "Debug";
    SpanExecutionType["Runtime"] = "Runtime";
})(SpanExecutionType || (SpanExecutionType = {}));
/** Whether the caller has permission to read this span's data. */
var SpanPermissionStatus;
(function (SpanPermissionStatus) {
    SpanPermissionStatus["Allow"] = "Allow";
    /** Some span fields are redacted due to permission constraints (e.g. attributes visible but payload hidden). */
    SpanPermissionStatus["PartialBlock"] = "PartialBlock";
    SpanPermissionStatus["Block"] = "Block";
})(SpanPermissionStatus || (SpanPermissionStatus = {}));
/** Storage provider that created or manages the attachment. */
var SpanAttachmentProvider;
(function (SpanAttachmentProvider) {
    SpanAttachmentProvider["Orchestrator"] = "Orchestrator";
    /** Span attachment stored by the observability platform. */
    SpanAttachmentProvider["LLMOps"] = "LLMOps";
})(SpanAttachmentProvider || (SpanAttachmentProvider = {}));
/** Whether the attachment is an input, output, or neither. */
var SpanAttachmentDirection;
(function (SpanAttachmentDirection) {
    SpanAttachmentDirection["None"] = "None";
    SpanAttachmentDirection["In"] = "In";
    SpanAttachmentDirection["Out"] = "Out";
})(SpanAttachmentDirection || (SpanAttachmentDirection = {}));

/** Maps integer Status values from the otel API to {@link SpanStatus} enum values. */
const SpanStatusMap = {
    0: SpanStatus.Unset,
    1: SpanStatus.Ok,
    2: SpanStatus.Error,
    3: SpanStatus.Running,
    4: SpanStatus.Restricted,
    5: SpanStatus.Cancelled,
};
/** Maps integer Source values from the otel API to {@link SpanSource} enum values. */
const SpanSourceMap = {
    0: SpanSource.Testing,
    1: SpanSource.Agents,
    2: SpanSource.ProcessOrchestration,
    3: SpanSource.ApiWorkflows,
    4: SpanSource.Robots,
    5: SpanSource.ConversationalAgentsService,
    6: SpanSource.IntegrationServiceTrigger,
    7: SpanSource.Playground,
    8: SpanSource.Governance,
    9: SpanSource.IXPUnstructuredAndComplexDocuments,
    10: SpanSource.CodedAgents,
    11: SpanSource.IXPCommunicationsMining,
    12: SpanSource.EnterpriseContextService,
    13: SpanSource.MCP,
    14: SpanSource.A2A,
    15: SpanSource.Serverless,
};
/** Maps integer VerbosityLevel values from the otel API to {@link SpanVerbosityLevel} enum values. */
const SpanVerbosityLevelMap = {
    0: SpanVerbosityLevel.Verbose,
    1: SpanVerbosityLevel.Trace,
    2: SpanVerbosityLevel.Information,
    3: SpanVerbosityLevel.Warning,
    4: SpanVerbosityLevel.Error,
    5: SpanVerbosityLevel.Critical,
    6: SpanVerbosityLevel.Off,
};
/** Maps integer ExecutionType values from the otel API to {@link SpanExecutionType} enum values. */
const SpanExecutionTypeMap = {
    0: SpanExecutionType.Debug,
    1: SpanExecutionType.Runtime,
};
/** Maps integer PermissionStatus values from the otel API to {@link SpanPermissionStatus} enum values. */
const SpanPermissionStatusMap = {
    0: SpanPermissionStatus.Allow,
    1: SpanPermissionStatus.PartialBlock,
    2: SpanPermissionStatus.Block,
};
/** Maps integer Provider values from the otel API to {@link SpanAttachmentProvider} enum values. */
const SpanAttachmentProviderMap = {
    0: SpanAttachmentProvider.Orchestrator,
    1: SpanAttachmentProvider.LLMOps,
};
/** Maps integer Direction values from the otel API to {@link SpanAttachmentDirection} enum values. */
const SpanAttachmentDirectionMap = {
    0: SpanAttachmentDirection.None,
    1: SpanAttachmentDirection.In,
    2: SpanAttachmentDirection.Out,
};

/**
 * Base path constants for different services
 */
const LLMOPS_BASE = 'llmopstenant_';
const INSIGHTS_RTM_BASE = 'insightsrtm_';

/**
 * Agent Traces Service Endpoints
 */
const AGENT_TRACES_ENDPOINTS = {
    /** Trace-level time-series of error counts grouped by error name. */
    GET_ERRORS_TIMELINE: `${INSIGHTS_RTM_BASE}/Traceview/errorsTimeline`,
    /** Trace-level time-series of latency (decimal seconds per series). */
    GET_LATENCY_TIMELINE: `${INSIGHTS_RTM_BASE}/Traceview/latencyTimeline`,
    /** Trace-level per-agent AGU/PLTU consumption totals. */
    GET_UNIT_CONSUMPTION: `${INSIGHTS_RTM_BASE}/Traceview/unitConsumption`,
    /** All spans for a single trace (flat array, not paginated). */
    GET_SPANS_BY_TRACE_ID: (traceId) => `${INSIGHTS_RTM_BASE}/Traceview/spans/${traceId}`,
    /** Paginated spans whose reference hierarchy contains the given reference id. */
    GET_SPANS_BY_REFERENCE: (referenceId) => `${INSIGHTS_RTM_BASE}/Traceview/spans/reference/${referenceId}`,
};

/**
 * Traces Service Endpoints
 */
const TRACES_ENDPOINTS = {
    /** GET spans for a trace (OTEL format). Query params: traceId, pageSize, agentId, isHistorical */
    GET_BY_TRACE_ID: `${LLMOPS_BASE}/api/Traces/v2/spans/otel`,
    /** POST specific spans by ID. traceId in query via params, spanIds array in body */
    POST_BY_IDS: `${LLMOPS_BASE}/api/Traces/v2/spans/otel/byIds`,
};

/**
 * SDK Telemetry constants.
 *
 * Only the SDK's identity (version, service name, role name, …) lives
 * here. The Application Insights connection string is injected into
 * `@uipath/core-telemetry` itself at publish time, and the generic attribute
 * keys (`Version`, `Service`, `CloudOrganizationName`, …) are owned by
 * `@uipath/core-telemetry` and consumed there — they are not part of the
 * SDK's public API.
 */
/** SDK version placeholder — patched by the SDK publish workflow. */
const CLOUD_ROLE_NAME = 'uipath-ts-sdk';

/**
 * UiPath TypeScript SDK Telemetry
 *
 * Constructs the SDK's own `TelemetryClient` and binds the SDK-local
 * `track` / `trackEvent` to it. Each consumer of `@uipath/core-telemetry`
 * does this independently, so events carry their own consumer's identity
 * and tenant context.
 */
// Keyed by `CLOUD_ROLE_NAME` so every SDK subpath bundle resolves to the
// same `TelemetryClient` instance at runtime. A single `initialize(...)`
// from the `UiPath` constructor therefore wires up `@track` decorators
// across every subpath bundle (`assets`, `feedback`, `tasks`, …).
const sdkClient = getOrCreateClient(CLOUD_ROLE_NAME);
const track = createTrack(sdkClient);
createTrackEvent(sdkClient);

class TracesService extends BaseService {
    transformOtelSpan(raw) {
        const { Attributes, ExpiryTimeUtc, Attachments, ...rest } = raw;
        const base = pascalToCamelCaseKeys(rest);
        return {
            ...base,
            attributes: Attributes,
            expiredTime: ExpiryTimeUtc,
            status: SpanStatusMap[raw.Status] ?? SpanStatus.Unset,
            source: raw.Source == null ? null : (SpanSourceMap[raw.Source] ?? null),
            verbosityLevel: raw.VerbosityLevel == null ? null : (SpanVerbosityLevelMap[raw.VerbosityLevel] ?? null),
            executionType: raw.ExecutionType == null ? null : (SpanExecutionTypeMap[raw.ExecutionType] ?? null),
            permissionStatus: raw.PermissionStatus == null ? null : (SpanPermissionStatusMap[raw.PermissionStatus] ?? null),
            attachments: Attachments ? Attachments.map(a => ({
                provider: SpanAttachmentProviderMap[a.Provider] ?? SpanAttachmentProvider.Orchestrator,
                id: a.Id,
                fileName: a.FileName,
                mimeType: a.MimeType,
                direction: SpanAttachmentDirectionMap[a.Direction] ?? SpanAttachmentDirection.None,
            })) : null,
        };
    }
    /**
     * Gets all spans for a specific trace ID.
     *
     * Returns up to `pageSize` spans (default 1000) in a single fetch.
     * Accepts both GUID format and OTEL 32-char hex format — the API normalizes both.
     *
     * @param traceId - Trace identifier
     * @param options - Optional filters {@link TracesGetByIdOptions}
     * @returns Promise resolving to an array of {@link SpanGetResponse}, each containing span identity, timing, status, source platform, attributes, verbosity, execution type, lineage context, and any file attachments.
     * @example
     * ```typescript
     * import { Traces } from '@uipath/uipath-typescript/traces';
     *
     * const traces = new Traces(sdk);
     * const spans = await traces.getById('<traceId>');
     * console.log(spans.length, spans[0].spanType, spans[0].status);
     * ```
     * @example
     * ```typescript
     * // Filter to a specific agent's spans
     * const agentSpans = await traces.getById('<traceId>', {
     *   agentId: '<agentId>',
     *   pageSize: 500,
     * });
     * ```
     */
    async getById(traceId, options) {
        if (!traceId)
            throw new ValidationError({ message: 'traceId is required for getById' });
        const { pageSize = 1000, agentId, includeExpiredSpans } = options ?? {};
        const params = { traceId, pageSize };
        if (agentId !== undefined)
            params.agentId = agentId;
        if (includeExpiredSpans !== undefined)
            params.isHistorical = includeExpiredSpans;
        const response = await this.get(TRACES_ENDPOINTS.GET_BY_TRACE_ID, { params });
        const spans = response.data?.Spans ?? [];
        return spans.map(span => this.transformOtelSpan(span));
    }
    /**
     * Gets specific spans by trace ID and span IDs.
     *
     * Accepts OTEL 16-char hex or GUID format for span IDs.
     *
     * @param traceId - Trace identifier
     * @param spanIds - List of span IDs to retrieve
     * @returns Promise resolving to an array of matching {@link SpanGetResponse}
     * @example
     * ```typescript
     * import { Traces } from '@uipath/uipath-typescript/traces';
     *
     * const traces = new Traces(sdk);
     *
     * // First retrieve all spans to find the IDs you want
     * const allSpans = await traces.getById('<traceId>');
     * const spanIds = allSpans.slice(0, 3).map(s => s.id);
     *
     * const subset = await traces.getSpansByIds('<traceId>', spanIds);
     * ```
     */
    async getSpansByIds(traceId, spanIds) {
        if (!traceId)
            throw new ValidationError({ message: 'traceId is required for getSpansByIds' });
        const response = await this.post(TRACES_ENDPOINTS.POST_BY_IDS, spanIds, { params: { traceId } });
        const spans = response.data ?? [];
        return spans.map(span => this.transformOtelSpan(span));
    }
}
__decorate([
    track('Traces.GetById')
], TracesService.prototype, "getById", null);
__decorate([
    track('Traces.GetSpansByIds')
], TracesService.prototype, "getSpansByIds", null);

/**
 * Maps a raw span record to a {@link AgentSpanGetResponse}
 */
const transformSpan = (span) => {
    const { expiryTimeUtc, ...rest } = span;
    return { ...rest, expiredTime: expiryTimeUtc };
};
/**
 * Service for retrieving UiPath Agent trace metrics.
 */
class AgentTracesService extends BaseService {
    /**
     * Retrieves a trace-level time-series of error counts grouped by error name.
     *
     * @param options - Optional window and filters
     * @returns Promise resolving to an array of {@link AgentTraceGetErrorsTimelineResponse}
     * @example
     * ```typescript
     * import { AgentTraces } from '@uipath/uipath-typescript/traces';
     *
     * const trace = new AgentTraces(sdk);
     *
     * // Get the errors timeline
     * const result = await trace.getErrorsTimeline();
     * result.forEach((point) => {
     *   console.log(`${point.date} ${point.name}: ${point.value} errors`);
     * });
     * ```
     * @example
     * ```typescript
     * import { AgentTraceExecutionType } from '@uipath/uipath-typescript/traces';
     *
     * // Get the errors timeline for an agent version within a time window
     * const filtered = await trace.getErrorsTimeline({
     *   startTime: new Date('2025-05-01T00:00:00Z'),
     *   endTime: new Date('2025-06-01T00:00:00Z'),
     *   agentId: '<agentId>',
     *   agentVersion: '1.0.0',
     *   executionType: AgentTraceExecutionType.Runtime,
     * });
     * ```
     */
    async getErrorsTimeline(options) {
        const response = await this.post(AGENT_TRACES_ENDPOINTS.GET_ERRORS_TIMELINE, this.buildTraceFilterBody(options));
        return response.data.data;
    }
    /**
     * Retrieves a trace-level time-series of latency.
     *
     * @param options - Optional window and filters
     * @returns Promise resolving to an array of {@link AgentTraceGetLatencyTimelineResponse}
     * @example
     * ```typescript
     * import { AgentTraces } from '@uipath/uipath-typescript/traces';
     *
     * const trace = new AgentTraces(sdk);
     *
     * // Get the latency timeline
     * const result = await trace.getLatencyTimeline();
     * result.forEach((point) => {
     *   console.log(`${point.date} ${point.name}: ${point.value}s`);
     * });
     * ```
     * @example
     * ```typescript
     * import { AgentTraceExecutionType } from '@uipath/uipath-typescript/traces';
     *
     * // Get the latency timeline for an agent version within a time window
     * const filtered = await trace.getLatencyTimeline({
     *   startTime: new Date('2025-05-01T00:00:00Z'),
     *   endTime: new Date('2025-06-01T00:00:00Z'),
     *   agentId: '<agentId>',
     *   agentVersion: '1.0.0',
     *   executionType: AgentTraceExecutionType.Runtime,
     * });
     * ```
     */
    async getLatencyTimeline(options) {
        const response = await this.post(AGENT_TRACES_ENDPOINTS.GET_LATENCY_TIMELINE, this.buildTraceFilterBody(options));
        return response.data.data;
    }
    /**
     * Retrieves trace-level per-agent unit consumption totals.
     *
     * @param options - Optional window and filters
     * @returns Promise resolving to an array of {@link AgentTraceGetUnitConsumptionResponse}
     * @example
     * ```typescript
     * import { AgentTraces } from '@uipath/uipath-typescript/traces';
     *
     * const trace = new AgentTraces(sdk);
     *
     * // Get per-agent unit consumption
     * const result = await trace.getUnitConsumption();
     * result.forEach((row) => {
     *   console.log(`${row.agentId}: ${row.agentUnitsConsumed} AGU, ${row.platformUnitsConsumed} PLTU`);
     * });
     * ```
     * @example
     * ```typescript
     * import { AgentTraceExecutionType } from '@uipath/uipath-typescript/traces';
     *
     * // Get per-agent unit consumption for an agent version within a time window
     * const filtered = await trace.getUnitConsumption({
     *   startTime: new Date('2025-05-01T00:00:00Z'),
     *   endTime: new Date('2025-06-01T00:00:00Z'),
     *   agentId: '<agentId>',
     *   agentVersion: '1.0.0',
     *   executionType: AgentTraceExecutionType.Runtime,
     * });
     * ```
     */
    async getUnitConsumption(options) {
        const response = await this.post(AGENT_TRACES_ENDPOINTS.GET_UNIT_CONSUMPTION, this.buildTraceFilterBody(options));
        return response.data.data;
    }
    /**
     * Retrieves every span belonging to a single trace.
     *
     * @param traceId - Identifier of the trace whose spans should be returned
     * @returns Promise resolving to an array of {@link AgentSpanGetResponse}
     * @example
     * ```typescript
     * import { AgentTraces } from '@uipath/uipath-typescript/traces';
     *
     * const trace = new AgentTraces(sdk);
     *
     * const spans = await trace.getSpansByTraceId('<traceId>');
     * spans.forEach((span) => {
     *   console.log(`${span.name} (${span.startTime} → ${span.endTime ?? 'in progress'})`);
     * });
     * ```
     */
    async getSpansByTraceId(traceId) {
        if (!traceId)
            throw new ValidationError({ message: 'traceId is required for getSpansByTraceId' });
        const response = await this.get(AGENT_TRACES_ENDPOINTS.GET_SPANS_BY_TRACE_ID(traceId));
        return (response.data ?? []).map(transformSpan);
    }
    /**
     * Retrieves spans whose reference hierarchy contains the given reference id.
     *
     * Returns a {@link PaginatedResponse} when pagination options (`pageSize`,
     * `cursor`, or `jumpToPage`) are provided, otherwise a
     * {@link NonPaginatedResponse}.
     *
     * @param referenceId - Reference id matched against each span's reference hierarchy
     * @param options - Optional pagination and hierarchy/time filters
     * @returns Promise resolving to a paginated or non-paginated list of {@link AgentSpanGetResponse}
     * @example
     * ```typescript
     * import { AgentTraces } from '@uipath/uipath-typescript/traces';
     *
     * const trace = new AgentTraces(sdk);
     *
     * // Get spans by referenceId
     * const result = await trace.getSpansByReference('<referenceId>');
     * result.items.forEach((span) => console.log(span.name));
     * ```
     * @example
     * ```typescript
     * import { AgentTraceExecutionType } from '@uipath/uipath-typescript/traces';
     *
     * // Get spans by referenceId within a trace and time window
     * const page = await trace.getSpansByReference('<referenceId>', {
     *   traceId: '<traceId>',
     *   executionType: AgentTraceExecutionType.Runtime,
     *   startTime: new Date('2025-05-01T00:00:00Z'),
     *   endTime: new Date('2025-06-01T00:00:00Z'),
     *   pageSize: 25,
     * });
     *
     * if (page.hasNextPage && page.nextCursor) {
     *   const next = await trace.getSpansByReference('<referenceId>', { cursor: page.nextCursor });
     * }
     * ```
     */
    async getSpansByReference(referenceId, options) {
        if (!referenceId)
            throw new ValidationError({ message: 'referenceId is required for getSpansByReference' });
        const { startTime, endTime, ...rest } = options ?? {};
        const apiOptions = {
            ...rest,
            ...(startTime !== undefined ? { startTime: startTime.toISOString() } : {}),
            ...(endTime !== undefined ? { endTime: endTime.toISOString() } : {}),
        };
        return PaginationHelpers.getAll({
            serviceAccess: this.createPaginationServiceAccess(),
            getEndpoint: () => AGENT_TRACES_ENDPOINTS.GET_SPANS_BY_REFERENCE(referenceId),
            method: HTTP_METHODS.GET,
            transformFn: transformSpan,
            excludeFromPrefix: Object.keys(apiOptions),
            pagination: {
                paginationType: PaginationType.OFFSET,
                itemsField: TRACEVIEW_SPANS_PAGINATION.ITEMS_FIELD,
                totalCountField: TRACEVIEW_SPANS_PAGINATION.TOTAL_COUNT_FIELD,
                paginationParams: {
                    pageSizeParam: AGENTS_OFFSET_PARAMS.PAGE_SIZE_PARAM,
                    offsetParam: AGENTS_OFFSET_PARAMS.OFFSET_PARAM,
                    countParam: AGENTS_OFFSET_PARAMS.COUNT_PARAM,
                    convertToSkip: false,
                    zeroBased: true,
                },
            },
        }, apiOptions);
    }
    buildTraceFilterBody(options) {
        const body = {};
        if (options?.startTime !== undefined)
            body.startTime = options.startTime.toISOString();
        if (options?.endTime !== undefined)
            body.endTime = options.endTime.toISOString();
        if (options?.folderKeys !== undefined)
            body.folderKeys = options.folderKeys;
        if (options?.agentId !== undefined)
            body.agentId = options.agentId;
        if (options?.agentVersion !== undefined)
            body.agentVersion = options.agentVersion;
        if (options?.executionType !== undefined)
            body.executionType = options.executionType;
        return body;
    }
}
__decorate([
    track('AgentTraces.GetErrorsTimeline')
], AgentTracesService.prototype, "getErrorsTimeline", null);
__decorate([
    track('AgentTraces.GetLatencyTimeline')
], AgentTracesService.prototype, "getLatencyTimeline", null);
__decorate([
    track('AgentTraces.GetUnitConsumption')
], AgentTracesService.prototype, "getUnitConsumption", null);
__decorate([
    track('AgentTraces.GetSpansByTraceId')
], AgentTracesService.prototype, "getSpansByTraceId", null);
__decorate([
    track('AgentTraces.GetSpansByReference')
], AgentTracesService.prototype, "getSpansByReference", null);

export { SpanExecutionType as AgentTraceExecutionType, AgentTracesService as AgentTraces, SpanAttachmentDirection, SpanAttachmentProvider, SpanExecutionType, SpanPermissionStatus, SpanSource, SpanStatus, SpanVerbosityLevel, TracesService as Traces };
