// @uipath/uipath-typescript/dist/document-understanding/index.cjs
'use strict';

// Auto-generated from the OpenAPI spec — do not edit manually.
exports.DocumentActionPriority = void 0;
(function (DocumentActionPriority) {
    DocumentActionPriority["Low"] = "Low";
    DocumentActionPriority["Medium"] = "Medium";
    DocumentActionPriority["High"] = "High";
    DocumentActionPriority["Critical"] = "Critical";
})(exports.DocumentActionPriority || (exports.DocumentActionPriority = {}));
exports.DocumentActionStatus = void 0;
(function (DocumentActionStatus) {
    DocumentActionStatus["Unassigned"] = "Unassigned";
    DocumentActionStatus["Pending"] = "Pending";
    DocumentActionStatus["Completed"] = "Completed";
})(exports.DocumentActionStatus || (exports.DocumentActionStatus = {}));
exports.DocumentActionType = void 0;
(function (DocumentActionType) {
    DocumentActionType["Validation"] = "Validation";
    DocumentActionType["Classification"] = "Classification";
})(exports.DocumentActionType || (exports.DocumentActionType = {}));

// Auto-generated from the OpenAPI spec — do not edit manually.
exports.ProjectProperties = void 0;
(function (ProjectProperties) {
    ProjectProperties["IsPredefined"] = "IsPredefined";
    ProjectProperties["SupportsTags"] = "SupportsTags";
    ProjectProperties["SupportsVersions"] = "SupportsVersions";
    ProjectProperties["IsSplittingEnabled"] = "IsSplittingEnabled";
})(exports.ProjectProperties || (exports.ProjectProperties = {}));
exports.ProjectType = void 0;
(function (ProjectType) {
    ProjectType["Classic"] = "Classic";
    ProjectType["Modern"] = "Modern";
    ProjectType["IXP"] = "IXP";
    ProjectType["Unknown"] = "Unknown";
})(exports.ProjectType || (exports.ProjectType = {}));
exports.ResourceStatus = void 0;
(function (ResourceStatus) {
    ResourceStatus["Unknown"] = "Unknown";
    ResourceStatus["NotStarted"] = "NotStarted";
    ResourceStatus["ExportInProgress"] = "ExportInProgress";
    ResourceStatus["ExportCompleted"] = "ExportCompleted";
    ResourceStatus["InProgress"] = "InProgress";
    ResourceStatus["Suspended"] = "Suspended";
    ResourceStatus["Resuming"] = "Resuming";
    ResourceStatus["Available"] = "Available";
    ResourceStatus["Error"] = "Error";
    ResourceStatus["Deleting"] = "Deleting";
})(exports.ResourceStatus || (exports.ResourceStatus = {}));
exports.ResourceType = void 0;
(function (ResourceType) {
    ResourceType["Specialized"] = "Specialized";
    ResourceType["Generative"] = "Generative";
    ResourceType["Custom"] = "Custom";
    ResourceType["Unknown"] = "Unknown";
})(exports.ResourceType || (exports.ResourceType = {}));

// Auto-generated from the OpenAPI spec — do not edit manually.
exports.MarkupType = void 0;
(function (MarkupType) {
    MarkupType["Unknown"] = "Unknown";
    MarkupType["Circled"] = "Circled";
    MarkupType["Underlined"] = "Underlined";
    MarkupType["Strikethrough"] = "Strikethrough";
})(exports.MarkupType || (exports.MarkupType = {}));
exports.ProcessingSource = void 0;
(function (ProcessingSource) {
    ProcessingSource["Unknown"] = "Unknown";
    ProcessingSource["Ocr"] = "Ocr";
    ProcessingSource["Pdf"] = "Pdf";
    ProcessingSource["PlainText"] = "PlainText";
    ProcessingSource["PdfAndOcr"] = "PdfAndOcr";
})(exports.ProcessingSource || (exports.ProcessingSource = {}));
exports.Rotation = void 0;
(function (Rotation) {
    Rotation["None"] = "None";
    Rotation["Rotated90"] = "Rotated90";
    Rotation["Rotated180"] = "Rotated180";
    Rotation["Rotated270"] = "Rotated270";
    Rotation["Other"] = "Other";
})(exports.Rotation || (exports.Rotation = {}));
exports.SectionType = void 0;
(function (SectionType) {
    SectionType["Vertical"] = "Vertical";
    SectionType["Paragraph"] = "Paragraph";
    SectionType["Header"] = "Header";
    SectionType["Footer"] = "Footer";
    SectionType["Table"] = "Table";
})(exports.SectionType || (exports.SectionType = {}));
exports.TextType = void 0;
(function (TextType) {
    TextType["Unknown"] = "Unknown";
    TextType["Text"] = "Text";
    TextType["Checkbox"] = "Checkbox";
    TextType["Handwriting"] = "Handwriting";
    TextType["Barcode"] = "Barcode";
    TextType["QRcode"] = "QRcode";
    TextType["Stamp"] = "Stamp";
    TextType["Logo"] = "Logo";
    TextType["Circle"] = "Circle";
    TextType["Underline"] = "Underline";
    TextType["Cut"] = "Cut";
})(exports.TextType || (exports.TextType = {}));
exports.WordGroupType = void 0;
(function (WordGroupType) {
    WordGroupType["Sentence"] = "Sentence";
    WordGroupType["TableCell"] = "TableCell";
    WordGroupType["TableRowEnd"] = "TableRowEnd";
    WordGroupType["Heading"] = "Heading";
    WordGroupType["Other"] = "Other";
})(exports.WordGroupType || (exports.WordGroupType = {}));

// Auto-generated from the OpenAPI spec — do not edit manually.
exports.ModelKind = void 0;
(function (ModelKind) {
    ModelKind["Extractor"] = "Extractor";
    ModelKind["Classifier"] = "Classifier";
})(exports.ModelKind || (exports.ModelKind = {}));
exports.ModelType = void 0;
(function (ModelType) {
    ModelType["IXP"] = "IXP";
    ModelType["Modern"] = "Modern";
    ModelType["Predefined"] = "Predefined";
})(exports.ModelType || (exports.ModelType = {}));

// Auto-generated from the OpenAPI spec — do not edit manually.
exports.ErrorSeverity = void 0;
(function (ErrorSeverity) {
    ErrorSeverity["Info"] = "Info";
    ErrorSeverity["Warning"] = "Warning";
    ErrorSeverity["Error"] = "Error";
})(exports.ErrorSeverity || (exports.ErrorSeverity = {}));

// Auto-generated from the OpenAPI spec — do not edit manually.
exports.ClassifierDocumentTypeType = void 0;
(function (ClassifierDocumentTypeType) {
    ClassifierDocumentTypeType["FormsAi"] = "FormsAi";
    ClassifierDocumentTypeType["SemiStructuredAi"] = "SemiStructuredAi";
    ClassifierDocumentTypeType["Helix"] = "Helix";
    ClassifierDocumentTypeType["Unknown"] = "Unknown";
})(exports.ClassifierDocumentTypeType || (exports.ClassifierDocumentTypeType = {}));
exports.CreateTaskPriority = void 0;
(function (CreateTaskPriority) {
    CreateTaskPriority["Low"] = "Low";
    CreateTaskPriority["Medium"] = "Medium";
    CreateTaskPriority["High"] = "High";
    CreateTaskPriority["Critical"] = "Critical";
})(exports.CreateTaskPriority || (exports.CreateTaskPriority = {}));
exports.GptFieldType = void 0;
(function (GptFieldType) {
    GptFieldType["Address"] = "Address";
    GptFieldType["Boolean"] = "Boolean";
    GptFieldType["Date"] = "Date";
    GptFieldType["Name"] = "Name";
    GptFieldType["Number"] = "Number";
    GptFieldType["Text"] = "Text";
})(exports.GptFieldType || (exports.GptFieldType = {}));
exports.JobStatus = void 0;
(function (JobStatus) {
    JobStatus["Succeeded"] = "Succeeded";
    JobStatus["Failed"] = "Failed";
    JobStatus["Running"] = "Running";
    JobStatus["NotStarted"] = "NotStarted";
})(exports.JobStatus || (exports.JobStatus = {}));
exports.ValidationDisplayMode = void 0;
(function (ValidationDisplayMode) {
    ValidationDisplayMode["Classic"] = "Classic";
    ValidationDisplayMode["Compact"] = "Compact";
})(exports.ValidationDisplayMode || (exports.ValidationDisplayMode = {}));

// Auto-generated from the OpenAPI spec — do not edit manually.
exports.ResultsDataSource = void 0;
(function (ResultsDataSource) {
    ResultsDataSource["Automatic"] = "Automatic";
    ResultsDataSource["Manual"] = "Manual";
    ResultsDataSource["ManuallyChanged"] = "ManuallyChanged";
    ResultsDataSource["Defaulted"] = "Defaulted";
    ResultsDataSource["External"] = "External";
})(exports.ResultsDataSource || (exports.ResultsDataSource = {}));

// Auto-generated from the OpenAPI spec — do not edit manually.
exports.ComparisonOperator = void 0;
(function (ComparisonOperator) {
    ComparisonOperator["Equals"] = "Equals";
    ComparisonOperator["NotEquals"] = "NotEquals";
    ComparisonOperator["Greater"] = "Greater";
    ComparisonOperator["Less"] = "Less";
    ComparisonOperator["GreaterOrEqual"] = "GreaterOrEqual";
    ComparisonOperator["LessOrEqual"] = "LessOrEqual";
})(exports.ComparisonOperator || (exports.ComparisonOperator = {}));
exports.Criticality = void 0;
(function (Criticality) {
    Criticality["Must"] = "Must";
    Criticality["Should"] = "Should";
})(exports.Criticality || (exports.Criticality = {}));
exports.FieldType = void 0;
(function (FieldType) {
    FieldType["Text"] = "Text";
    FieldType["Number"] = "Number";
    FieldType["Date"] = "Date";
    FieldType["Name"] = "Name";
    FieldType["Address"] = "Address";
    FieldType["Keyword"] = "Keyword";
    FieldType["Set"] = "Set";
    FieldType["Boolean"] = "Boolean";
    FieldType["Table"] = "Table";
    FieldType["Internal"] = "Internal";
    FieldType["FieldGroup"] = "FieldGroup";
    FieldType["MonetaryQuantity"] = "MonetaryQuantity";
})(exports.FieldType || (exports.FieldType = {}));
exports.LogicalOperator = void 0;
(function (LogicalOperator) {
    LogicalOperator["AND"] = "AND";
    LogicalOperator["OR"] = "OR";
})(exports.LogicalOperator || (exports.LogicalOperator = {}));
exports.RuleType = void 0;
(function (RuleType) {
    RuleType["Mandatory"] = "Mandatory";
    RuleType["PossibleValues"] = "PossibleValues";
    RuleType["Regex"] = "Regex";
    RuleType["StartsWith"] = "StartsWith";
    RuleType["EndsWith"] = "EndsWith";
    RuleType["FixedLength"] = "FixedLength";
    RuleType["IsNumeric"] = "IsNumeric";
    RuleType["IsDate"] = "IsDate";
    RuleType["IsEmail"] = "IsEmail";
    RuleType["Contains"] = "Contains";
    RuleType["Expression"] = "Expression";
    RuleType["TableExpression"] = "TableExpression";
    RuleType["IsEmpty"] = "IsEmpty";
})(exports.RuleType || (exports.RuleType = {}));

// Auto-generated from the OpenAPI spec — do not edit manually.
exports.ActionStatus = void 0;
(function (ActionStatus) {
    ActionStatus["Unassigned"] = "Unassigned";
    ActionStatus["Pending"] = "Pending";
    ActionStatus["Completed"] = "Completed";
})(exports.ActionStatus || (exports.ActionStatus = {}));

var index = /*#__PURE__*/Object.freeze({
    __proto__: null,
    get ActionStatus () { return exports.ActionStatus; },
    get ClassifierDocumentTypeType () { return exports.ClassifierDocumentTypeType; },
    get ComparisonOperator () { return exports.ComparisonOperator; },
    get CreateTaskPriority () { return exports.CreateTaskPriority; },
    get Criticality () { return exports.Criticality; },
    get DocumentActionPriority () { return exports.DocumentActionPriority; },
    get DocumentActionStatus () { return exports.DocumentActionStatus; },
    get DocumentActionType () { return exports.DocumentActionType; },
    get ErrorSeverity () { return exports.ErrorSeverity; },
    get FieldType () { return exports.FieldType; },
    get GptFieldType () { return exports.GptFieldType; },
    get JobStatus () { return exports.JobStatus; },
    get LogicalOperator () { return exports.LogicalOperator; },
    get MarkupType () { return exports.MarkupType; },
    get ModelKind () { return exports.ModelKind; },
    get ModelType () { return exports.ModelType; },
    get ProcessingSource () { return exports.ProcessingSource; },
    get ProjectProperties () { return exports.ProjectProperties; },
    get ProjectType () { return exports.ProjectType; },
    get ResourceStatus () { return exports.ResourceStatus; },
    get ResourceType () { return exports.ResourceType; },
    get ResultsDataSource () { return exports.ResultsDataSource; },
    get Rotation () { return exports.Rotation; },
    get RuleType () { return exports.RuleType; },
    get SectionType () { return exports.SectionType; },
    get TextType () { return exports.TextType; },
    get ValidationDisplayMode () { return exports.ValidationDisplayMode; },
    get WordGroupType () { return exports.WordGroupType; }
});

exports.DuFramework = index;
