// @uipath/core-telemetry/dist/index.mjs
import { BatchLogRecordProcessor, LoggerProvider } from '@opentelemetry/sdk-logs';

/**
 * Common telemetry attribute keys + Application Insights connection string.
 *
 * Producer-specific identity (SDK version, role name, etc.) is supplied at
 * runtime via `TelemetryClient.initialize(...)` so the same package can be
 * reused by multiple packages. The Application Insights connection string is
 * shared across all consumers and is patched into this file at publish time.
 */
// Connection string placeholder that will be replaced during build
const CONNECTION_STRING = 'InstrumentationKey=a6efa11d-1feb-4508-9738-e13e12dcae5e;IngestionEndpoint=https://westeurope-5.in.applicationinsights.azure.com/;LiveEndpoint=https://westeurope.livediagnostics.monitor.azure.com/;ApplicationId=7c58eb1c-9581-4ba6-839e-11725848a037';
const VERSION = 'Version';
const SERVICE = 'Service';
const CLOUD_ORGANIZATION_NAME = 'CloudOrganizationName';
const CLOUD_TENANT_NAME = 'CloudTenantName';
const CLOUD_ORGANIZATION_ID = 'CloudOrganizationId';
const CLOUD_TENANT_ID = 'CloudTenantId';
const CLOUD_USER_ID = 'CloudUserId';
const CLOUD_URL = 'CloudUrl';
const CLOUD_CLIENT_ID = 'CloudClientId';
const CLOUD_REDIRECT_URI = 'CloudRedirectUri';
const APP_NAME = 'ApplicationName';
/** Default value used when an attribute has no resolved value. */
const UNKNOWN = '';

const INSTRUMENTATION_KEY_RE = /InstrumentationKey=([^;]+)/;
const INGESTION_ENDPOINT_RE = /IngestionEndpoint=([^;]+)/;
/**
 * Sends every emitted log record as an Application Insights custom event.
 * Plugged into OpenTelemetry's `BatchLogRecordProcessor`, which buffers and
 * flushes log records to this exporter on its own schedule.
 */
class ApplicationInsightsEventExporter {
    constructor(connectionString, tags) {
        this.connectionString = connectionString;
        this.tags = tags;
    }
    export(logs, resultCallback) {
        try {
            for (const log of logs) {
                this.sendAsCustomEvent(log);
            }
            resultCallback({ code: 0 });
        }
        catch (error) {
            const err = error instanceof Error ? error : new Error(String(error));
            console.debug('Failed to export logs to Application Insights:', err);
            resultCallback({ code: 1, error: err });
        }
    }
    shutdown() {
        return Promise.resolve();
    }
    sendAsCustomEvent(logRecord) {
        const eventName = String(logRecord.body);
        const payload = {
            name: 'Microsoft.ApplicationInsights.Event',
            time: new Date().toISOString(),
            iKey: this.extractInstrumentationKey(),
            data: {
                baseType: 'EventData',
                baseData: {
                    ver: 2,
                    name: eventName,
                    properties: this.convertAttributesToProperties(logRecord.attributes),
                },
            },
            tags: {
                'ai.cloud.role': this.tags.cloudRoleName,
                'ai.cloud.roleInstance': this.tags.cloudRoleInstance,
            },
        };
        void this.sendToApplicationInsights(payload);
    }
    extractInstrumentationKey() {
        const match = INSTRUMENTATION_KEY_RE.exec(this.connectionString);
        return match ? match[1] : '';
    }
    convertAttributesToProperties(attributes) {
        const properties = {};
        for (const [key, value] of Object.entries(attributes ?? {})) {
            properties[key] = String(value);
        }
        return properties;
    }
    async sendToApplicationInsights(payload) {
        try {
            const ingestionEndpoint = this.extractIngestionEndpoint();
            if (!ingestionEndpoint) {
                console.debug('No ingestion endpoint found in connection string');
                return;
            }
            const url = `${ingestionEndpoint}/v2/track`;
            const response = await fetch(url, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(payload),
            });
            if (!response.ok) {
                console.debug(`Failed to send event telemetry: ${response.status} ${response.statusText}`);
            }
        }
        catch (error) {
            console.debug('Error sending event telemetry to Application Insights:', error);
        }
    }
    extractIngestionEndpoint() {
        const match = INGESTION_ENDPOINT_RE.exec(this.connectionString);
        return match ? match[1] : '';
    }
}
/**
 * Telemetry client owned by a single consumer (e.g. the SDK or the Coded
 * Action Apps package). Each consumer instantiates its own client so that
 * its identity (`cloudRoleName`, `serviceName`, `sdkVersion`, …) and tenant
 * context flow through to its own `Logger` and exporter pipeline. Two
 * consumers running in the same process emit independent events — they
 * share the Application Insights connection string but nothing else.
 *
 * Records are emitted via `Logger.emit` and batched by
 * `BatchLogRecordProcessor` before being handed to the Application Insights
 * exporter.
 */
class TelemetryClient {
    constructor() {
        this.isInitialized = false;
    }
    initialize(options) {
        if (this.isInitialized) {
            console.debug('Telemetry client has already been initialized');
            return;
        }
        this.isInitialized = true;
        this.options = options;
        this.telemetryContext = options.context;
        try {
            if (!this.isValidConnectionString(CONNECTION_STRING)) {
                return;
            }
            this.setupTelemetryProvider(CONNECTION_STRING);
        }
        catch (error) {
            console.debug('Failed to initialize telemetry:', error);
        }
    }
    track(eventName, name, extraAttributes = {}) {
        try {
            if (!this.logger) {
                return;
            }
            const finalDisplayName = name ?? eventName;
            const attributes = this.getEnrichedAttributes(extraAttributes, eventName);
            this.logger.emit({
                body: finalDisplayName,
                attributes,
                timestamp: Date.now(),
            });
        }
        catch (error) {
            console.debug('Failed to track telemetry event:', error);
        }
    }
    /**
     * Default event name (e.g. `Sdk.Run`) used when a tracker fires without
     * an explicit display name. Returns `undefined` until `initialize` runs.
     */
    getDefaultEventName() {
        return this.options?.defaultEventName;
    }
    /**
     * Sets the authenticated user's id, reported as `CloudUserId` on every
     * subsequently emitted event. Empty values are ignored and the
     * previously set user id, if any, is kept.
     */
    setUserId(userId) {
        if (userId) {
            this.cloudUserId = userId;
        }
    }
    setupTelemetryProvider(connectionString) {
        // `setupTelemetryProvider` is only called from `initialize` after
        // `this.options` has been assigned, so the non-null assertion is safe.
        const opts = this.options;
        const exporter = new ApplicationInsightsEventExporter(connectionString, {
            cloudRoleName: opts.cloudRoleName,
            cloudRoleInstance: opts.sdkVersion,
        });
        const processor = new BatchLogRecordProcessor(exporter);
        this.logProvider = new LoggerProvider({
            processors: [processor],
        });
        this.logger = this.logProvider.getLogger(opts.loggerName);
    }
    isValidConnectionString(connectionString) {
        // Build placeholders are emitted as `$CONNECTION_STRING` literally
        // until the publish workflow patches them. Treat any unsubstituted
        // placeholder as "no connection string available".
        return Boolean(connectionString) && !connectionString.startsWith('$');
    }
    getEnrichedAttributes(extraAttributes, eventName) {
        const opts = this.options;
        return {
            [APP_NAME]: opts?.serviceName ?? UNKNOWN,
            [VERSION]: opts?.sdkVersion ?? UNKNOWN,
            [SERVICE]: eventName,
            [CLOUD_URL]: this.createCloudUrl(),
            [CLOUD_ORGANIZATION_NAME]: this.telemetryContext?.orgName ?? UNKNOWN,
            [CLOUD_TENANT_NAME]: this.telemetryContext?.tenantName ?? UNKNOWN,
            [CLOUD_ORGANIZATION_ID]: this.telemetryContext?.orgId ?? UNKNOWN,
            [CLOUD_TENANT_ID]: this.telemetryContext?.tenantId ?? UNKNOWN,
            [CLOUD_USER_ID]: this.cloudUserId ?? UNKNOWN,
            [CLOUD_REDIRECT_URI]: this.telemetryContext?.redirectUri ?? UNKNOWN,
            [CLOUD_CLIENT_ID]: this.telemetryContext?.clientId ?? UNKNOWN,
            ...extraAttributes,
        };
    }
    createCloudUrl() {
        const baseUrl = this.telemetryContext?.baseUrl;
        // Prefer the org/tenant ids; fall back to the names when absent.
        const org = this.telemetryContext?.orgId || this.telemetryContext?.orgName;
        const tenant = this.telemetryContext?.tenantId || this.telemetryContext?.tenantName;
        if (!baseUrl || !org || !tenant) {
            return UNKNOWN;
        }
        return `${baseUrl}/${org}/${tenant}`;
    }
}

/**
 * Each consumer ships as multiple subpath bundles and Rollup inlines this
 * shim into every one — `new TelemetryClient()` would hand each bundle its
 * own (uninitialized) client. `getOrCreateClient(name)` instead parks the
 * client on `globalThis` so every bundle shares one client per name.
 */
const REGISTRY_KEY = Symbol.for('@uipath/core-telemetry/clients');
function getGlobalRegistry() {
    const holder = globalThis;
    let registry = holder[REGISTRY_KEY];
    if (!registry) {
        registry = {};
        holder[REGISTRY_KEY] = registry;
    }
    return registry;
}
/**
 * Returns the named `TelemetryClient`, creating it on first access.
 *
 * @param name Stable identifier for the consumer (e.g. `"uipath-ts-sdk"`,
 *   `"uipath-ts-coded-action-apps"`). Subpath bundles within the same
 *   consumer share a client by passing the same name.
 */
function getOrCreateClient(name) {
    const registry = getGlobalRegistry();
    if (!registry[name]) {
        registry[name] = new TelemetryClient();
    }
    return registry[name];
}

/**
 * `@track` decorator and `trackEvent` function factories.
 */
function createTrackedFunction(originalFunction, { client, nameOrOptions, fallbackName, opts }) {
    const wrapped = function (...args) {
        let shouldTrack = true;
        if (opts.condition !== undefined) {
            shouldTrack =
                typeof opts.condition === 'function'
                    ? opts.condition.apply(this, args)
                    : opts.condition;
        }
        if (shouldTrack) {
            const serviceMethod = typeof nameOrOptions === 'string' ? nameOrOptions : fallbackName;
            client.track(serviceMethod, client.getDefaultEventName(), opts.attributes);
        }
        return originalFunction.apply(this, args);
    };
    return wrapped;
}
/**
 * Factory: returns a `track` decorator bound to a specific
 * `TelemetryClient`. Each consumer creates its own decorator from its own
 * client so events carry that consumer's identity.
 *
 * @example
 * ```ts
 * const sdkClient = new TelemetryClient();
 * sdkClient.initialize({ ... });
 * export const track = createTrack(sdkClient);
 *
 * // Consumer code
 * @track('Queue.GetAll')
 * async getAll() { ... }
 *
 * @track('Assets.Update', { condition: false })
 * async update() { ... }
 * ```
 */
function createTrack(client) {
    function trackFactory(nameOrOptions, options) {
        const opts = typeof nameOrOptions === 'object' ? nameOrOptions : options ?? {};
        function decoratorImpl(target, propertyKey, descriptor) {
            if (descriptor && typeof descriptor.value === 'function') {
                const original = descriptor.value;
                descriptor.value = createTrackedFunction(original, {
                    client,
                    nameOrOptions,
                    fallbackName: typeof propertyKey === 'string'
                        ? propertyKey
                        : 'unknown_method',
                    opts,
                });
                return descriptor;
            }
            const fn = target;
            return createTrackedFunction(fn, {
                client,
                nameOrOptions,
                fallbackName: fn.name || 'unknown_function',
                opts,
            });
        }
        return decoratorImpl;
    }
    return trackFactory;
}
/**
 * Factory: returns a `trackEvent` function bound to a specific
 * `TelemetryClient`. Use for ad-hoc tracking outside method boundaries.
 *
 * @example
 * ```ts
 * const trackEvent = createTrackEvent(sdkClient);
 * trackEvent('Sdk.Auth');
 * ```
 */
function createTrackEvent(client) {
    return function trackEvent(eventName, name, attributes) {
        client.track(eventName, name, attributes);
    };
}

export { TelemetryClient, createTrack, createTrackEvent, getOrCreateClient };
